---
title: "Promoter profiles"
output: html_document
editor_options: 
  chunk_output_type: console
---


```{r setup, include=FALSE}
options(stringsAsFactors = FALSE)
knitr::opts_chunk$set(echo = TRUE)
library(GenomicRanges)
library(rtracklayer)
library(tidyverse)
library(ComplexHeatmap)
library(rslurm)

source("../../util/plotting_functions.R")
source("../../util/intersect_functions.R")
source("../../util/_setup.R")
```

```{r import}
gencode_gr <- rtracklayer::import("/scratch/Shares/rinnclass/data/gencode.v32.annotation.gtf")

fl <- list.files("/scratch/Shares/rinnclass/data/consensus_peaks", full.names = TRUE)
consensus_peaks <- lapply(fl, rtracklayer::import)
names(consensus_peaks) <- gsub("/scratch/Shares/rinnclass/data/consensus_peaks/|.bed", "", fl)

filtered_consensus_peaks <- consensus_peaks[sapply(consensus_peaks, length) > 250]

peak_occurrence_df <- read_csv("../07_conservative_reservoirs/results/peak_occurrence_df.csv")
```

```{r make-promoters}
genes <- gencode_gr[gencode_gr$type == "gene"]
all_promoters_gr <- promoters(genes, upstream = 3e3, downstream = 3e3)
```

```{r rslurm-promoter-profile}
# We don't want to include the histone marks or pol2
to_filter <- c("H3K27ac", "H3K36me3", "H3K4me1", "H3K4me2", "H3K4me3", "H3K9ac", "H3K9me3", 
               "H4K20me1", "POLR2A", "POLR2AphosphoS2", "POLR2AphosphoS5")
peak_list <- filtered_consensus_peaks[!(names(filtered_consensus_peaks) %in% to_filter)]


generate_promoter_profile <- function(gene_id) {
  promoter <- all_promoters_gr[all_promoters_gr$gene_id == gene_id]
  promoter_matrix <- make_promoter_binding_matrix(peak_list, promoter)
  promoter_profile <- t(as.matrix(colSums(promoter_matrix)))
  rownames(promoter_profile) <- gene_id
  return(promoter_profile)
}


hmm <- generate_promoter_profile("ENSG00000000971.15")
hmm[1:10]
plot(hmm[1,])
pars <- peak_occurrence_df %>% filter(number_of_dbp > 0) %>% dplyr::select(gene_id) %>% dplyr::slice(1:40)

pp_matrix <- matrix(numeric(), ncol = 6000)
for(i in 1:nrow(pars)) {
  print(pars$gene_id[[i]])
  pp <- generate_promoter_profile(pars$gene_id[[i]])
  pp_matrix <- rbind(pp_matrix, pp)
}
pheatmap::pheatmap(pp_matrix, cluster_cols = FALSE)



sjob <- slurm_apply(generate_promoter_profile, pars, jobname = 'promoter_profiles',
                      add_objects = c("all_promoters_gr", "peak_list", "make_promoter_binding_matrix", "extract_peak_view"),
                      nodes = 5, cpus_per_node = 10, 
                      slurm_options = list(time = '20:00:00'),
                      submit = FALSE)
```

```{r promoter-profile-results}
# Gather results
res <- read_rds("_rslurm_promoter_profiles/results_0.RDS")
res
res_matrix <- do.call(rbind, res)
```
