---
title: '01'
author: "James Pratt"
date: "3/12/2021"
output: html_document
editor_options: 
  chunk_output_type: console
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
knitr::opts_chunk$set(echo = TRUE)
library(tidyverse)
library(GenomicRanges)
source("util/intersect_functions.R")
source("util/plotting_functions.R")
source("util/_setup.R")
library(Gviz)
library(ggplot2)
```

## R Markdown

This is an R Markdown document. Markdown is a simple formatting syntax for authoring HTML, PDF, and MS Word documents. For more details on using R Markdown see <http://rmarkdown.rstudio.com>.

When you click the **Knit** button a document will be generated that includes both content as well as the output of any embedded R code chunks within the document. You can embed an R code chunk like this:

```{r}
lncrna_genebody <- rtracklayer::import("/scratch/Shares/rinnclass/James/CLASS_2021/analysis/lncrna_genebody.gtf")
lncrna_mrna_genebody <- rtracklayer::import("/scratch/Shares/rinnclass/James/CLASS_2021/analysis/lncrna_mrna_genebody.gtf")
lncrna_mrna_promoters <- rtracklayer::import("/scratch/Shares/rinnclass/James/CLASS_2021/analysis/lncrna_mrna_promoters.gtf")
lncrna_promoters <- rtracklayer::import("/scratch/Shares/rinnclass/James/CLASS_2021/analysis/lncrna_promoters.gtf")
mrna_promoters <- rtracklayer::import("/scratch/Shares/rinnclass/James/CLASS_2021/analysis/mrna_promoters.gtf")
mrna_genebody <- rtracklayer::import("/scratch/Shares/rinnclass/James/CLASS_2021/analysis/mrna_genebody.gtf")
```

```{r}
promoter_peak_occurence <- count_peaks_per_feature(lncrna_mrna_promoters, filtered_consensus_peaks, 
                                                   type = "occurrence")


## or just read in peak_occurence_dataframe.csv
peak_occurence_df <- read_csv("/scratch/Shares/rinnclass/James/CLASS_2021/analysis/01_global_peak_properties/results/peak_occurence_dataframe.csv")
```
```{r}
g <- ggplot(peak_occurence_df, aes(x = number_of_dbp))
 g + geom_density(alpha = 0.2, color = "#424242", fill = "#424242") +
     theme_paperwhite() +
     xlab(expression("Number of DBPs")) +
     ylab(expression("Density")) +
     ggtitle("Promoter binding events",
             subtitle = "mRNA and lncRNA genes")
```

```{r}
##filtering the dataframe
lncrna_peak_occurence_df <- filter(peak_occurence_df, gene_type == "lncRNA")
mrna_peak_occurence_df <- filter(peak_occurence_df, gene_type == "protein_coding")

## Including Plots

You can also embed plots, for example:

```{r pressure, echo=FALSE}
plot(pressure)
```

Note that the `echo = FALSE` parameter was added to the code chunk to prevent printing of the R code that generated the plot.
