---
title: "14-2: RNA-seq differential expression"
output: html_document
editor_options: 
  chunk_output_type: console
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
options(stringsAsFactors = FALSE)
library(tidyverse)
library(tximport)
library(DESeq2)

source("util/plotting_functions.R")
source("util/_setup.R")
```

Read in the count data.

```{r}
samplesheet <- read_csv("rnaseq/samplesheet.csv")

# We want to select just a subset of these samples
# so that we can run differential expression for two conditions
samplesheet <- samplesheet %>%
  filter(condition %in% c("cytosolic_fraction", "nuclear_fraction"))
```

```{r}
# We want to include the gene annotation info in the DEseq2 object,
# so we'll need to read that in.
gencode_gtf <- rtracklayer::import("/scratch/Shares/rinnclass/data/genomes/gencode.v32.annotation.gtf")
genes <- gencode_gtf[gencode_gtf$type == "gene"]

# Salmon's primary level of quantifying is at the transcript-level.
# Since we're doing differential expression at the gene-level
# Salmon needs to be able to map transcripts to genes.
tx2gene <- gencode_gtf[gencode_gtf$type == "transcript"] %>% 
  as.data.frame() %>%
  dplyr::select(transcript_id, gene_id)

# We'll also create an object that maps gene IDs to gene symbols
# This will come in handy in looking at our results
g2s <- genes %>% as.data.frame() %>%
  dplyr::select(gene_id, gene_name)
```

```{r}
# Read in the count data from salmon
# For DESeq2 we want the raw counts
# The easiest way to import it in the proper format is to 
# use the tximport package
files <- file.path("rnaseq/results/star_salmon",
                   samplesheet$sample_name, 
                   "quant.sf")
names(files) <- samplesheet$sample_name

txi <- tximport(files, type = c("salmon"), tx2gene = tx2gene)
```

```{r}
# Will be adding the gene information, the count data, and the sample information
# to the DESeq function. We want to ensure that all of this data is synced up
# So we need to reorder the data so that it's all matching.
names(genes) <- genes$gene_id
genes <- genes[rownames(txi$counts)]
stopifnot(all(genes$gene_id == rownames(txi$counts)))

# Create rownames for the sample sheet and then use that to re-order
samplesheet <- samplesheet %>%
  mutate(row_names = sample_name) %>%
  column_to_rownames("row_names")
samplesheet <- samplesheet[colnames(txi$counts),]
stopifnot(all(rownames(samplesheet) == colnames(txi$counts)))
```

The DESeq2 function requires gene information in GRanges format, the count data, and the sample information.

```{r}
# The conditions that we'll be comparing here are 
# are nuclear and cytoplasmic fractions
# Normally with WT and knockout experiments, you'd
# compare the treatment (KO) condition back to the control
# condition (WT)
# Here we don't really have an obvious choice of "control"
# So let's just choose to compare back to nuclear as our control
# To do that we make it the first level of the factor
samplesheet$condition <- factor(samplesheet$condition, levels = c("nuclear_fraction", "cytosolic_fraction"))

# Now we have all the components that we need to add to the 
# DESeq2 function. So now let's create our DEseq2 object
set.seed(789723)
dds <- DESeqDataSetFromTximport(txi,
                                design = ~ condition,
                                colData = samplesheet,
                                rowData = genes)
```

```{r}
# We now have the object set up to 
# run the differential expression analysis
# and it's a very simple command to do so. 
# Even though it's doing a lot of complicated statistics!
dds <- DESeq(dds)
```

```{r}
# The first thing that we'll want to do is make a PCA plot.
rlog_counts <- rlog(dds)

rld_pca <- prcomp(t(assay(rlog_counts)))
rld_prcomps <- rld_pca$x %>% as.data.frame() %>%
  rownames_to_column("sample_name") %>%
  select(sample_name, PC1, PC2)
rld_prcomps <- merge(samplesheet, rld_prcomps)
g <- ggplot(rld_prcomps, aes(x = PC1, y = PC2, color = condition))
g + geom_point() 
```

```{r}
# Now we can check out the results from this analysis.
resultsNames(dds)
res <- results(dds, name = "condition_cytosolic_fraction_vs_nuclear_fraction")

# It's easier to view it as a data.frame so we'll convert it.
res_df <- res %>% as.data.frame() %>%
  rownames_to_column("gene_id") %>%
  merge(g2s)
```

```{r}
# We can now make some standard plots to check out these results

# Volcano plot
ggplot(res_df, aes(x = log2FoldChange, y = -log10(padj))) + 
  geom_point()

# MA plot
ggplot(res_df, aes(x = log10(baseMean), y = log2FoldChange)) + 
  geom_point()

# P-value histogram
hist(res_df$padj)
```

```{r}
# Now on your own, find the top 10 genes that are nuclear vs. cytoplasmic.
# There's lots more to explore with this type of data and these results
# and we're opening to exploring more if you guys would like!
# This is just the very basics.
```


# DESeq with all subcellular fractions

```{r}
samplesheet <- read_csv("rnaseq/samplesheet.csv")

# We of course need to swap the labels for total and insoluble cytoplasmic
# From looking in the browser at genes that should be nuclear: NEAT1 for ex.
# We have concluded that 
# hepg2_R2 -- is whole cell / total
# hepg2_insoluble_cytoplasmic_fraction_R2 -- is whole cell / total
# hepg2_R1 -- is insoluble_cytoplasmic
# hepg2_insoluble_cytoplasmic_fraction_R1 -- is insoluble_cytoplasmic

# The cleanest thing to do would be to fix it on the design file and re-run the nf-core pipeline,
# but for now we can swap it here.
samplesheet[which(samplesheet$sample_name == "hepg2_R1"), "condition"] <- "insoluble_cytoplasmic_fraction"
samplesheet[which(samplesheet$sample_name == "hepg2_insoluble_cytoplasmic_fraction_R2"), "condition"] <- "total"
```


```{r}
# We now won't filter the sample sheet and we'll read in all the data
files <- file.path("rnaseq/results/star_salmon",
                   samplesheet$sample_name, 
                   "quant.sf")
names(files) <- samplesheet$sample_name

txi <- tximport(files, type = c("salmon"), tx2gene = tx2gene)
```


```{r}
# We'll do the same housekeeping as before to make sure the genes and the samples are in the same order as
# the count data.
names(genes) <- genes$gene_id
genes <- genes[rownames(txi$counts)]
stopifnot(all(genes$gene_id == rownames(txi$counts)))

# Create rownames for the sample sheet and then use that to re-order
samplesheet <- samplesheet %>%
  mutate(row_names = sample_name) %>%
  column_to_rownames("row_names")
samplesheet <- samplesheet[colnames(txi$counts),]
stopifnot(all(rownames(samplesheet) == colnames(txi$counts)))
```

```{r}
# In order to make the total (whole cell) condition be the one that everything is compared back to
# we will want to set the factor levels with "total" first
samplesheet$condition <- factor(samplesheet$condition, levels = c("total", "membrane_fraction", 
                                                                  "insoluble_cytoplasmic_fraction", 
                                                                  "cytosolic_fraction", "nuclear_fraction"))

dds <- DESeqDataSetFromTximport(txi,
                                design = ~ condition,
                                colData = samplesheet,
                                rowData = genes)

# Run the DESeq stats
dds <- DESeq(dds)
```

```{r}
# We now have a bunch more results
resultsNames(dds)

# Let's just look at one of the results


res <- results(dds, name = "condition_membrane_fraction_vs_total")

res_shrunken <- lfcShrink(dds, coef = "condition_membrane_fraction_vs_total",  res = res)


# Let's quickly compare the unshrunken vs shrunken fold changes.
res_df <- res %>% as.data.frame() %>%
  rownames_to_column("gene_id") %>%
  merge(g2s) %>%
  mutate(result_name = "condition_membrane_fraction_vs_total")

# Set the x-axes the same.
summary(res_df$log2FoldChange)

ggplot(res_df, aes(x = log2FoldChange, y = -log10(padj))) + 
  geom_point() + xlim(-24, 22)


res_shrunken_df <- res_shrunken %>% as.data.frame() %>%
  rownames_to_column("gene_id") %>%
  merge(g2s) %>%
  mutate(result_name = "condition_membrane_fraction_vs_total")

ggplot(res_shrunken_df, aes(x = log2FoldChange, y = -log10(padj))) + 
  geom_point() + xlim(-24, 22)
```

```{r}
# Okay, now that we have results for each comparison back to total,
# we can combine the fold changes into a matrix and make a heatmap / cluster the genes
# By which cellular fraction they are enriched in.
# A good starting point would be to use a for loop to make a data.frame with all the results
# and then you can make that into a matrix with pivot_wider.
results_names <- resultsNames(dds)
# We don't care about the intercept, so we can leave that out
results_names <- results_names[-1]

res_df <- data.frame("gene_id" = character(), 
                     "baseMean" = numeric(), 
                     "log2FoldChange" = numeric(), 
                     "lfcSE" = numeric(),
                     "stat" = numeric(),
                     "pvalue" = numeric(),
                     "padj" = numeric(),
                     "gene_name" = character(),
                     "result_name" = character())

res_shrunken_df <- data.frame("gene_id" = character(), 
                              "baseMean" = numeric(), 
                              "log2FoldChange" = numeric(), 
                              "lfcSE" = numeric(),
                              "stat" = numeric(),
                              "pvalue" = numeric(),
                              "padj" = numeric(),
                              "gene_name" = character(),
                              "result_name" = character())


for(i in 1:length(results_names)) {
  results_name <- results_names[i]
  res <- results(dds, name = results_name)
  res_shrunken <- lfcShrink(dds, coef = results_name,  res = res)
  
  tmp_res_df <- res %>% as.data.frame() %>%
    rownames_to_column("gene_id") %>%
    merge(g2s) %>%
    mutate(result_name = results_name)
  
  
  tmp_res_shrunken_df <- res_shrunken %>% as.data.frame() %>%
    rownames_to_column("gene_id") %>%
    merge(g2s) %>%
    mutate(result_name = results_name)
  
  # Append to full data.frame
  res_df <- bind_rows(res_df, tmp_res_df)
  res_shrunken_df <- bind_rows(res_shrunken_df, tmp_res_shrunken_df)
}

# Let's move forward with the shrunken log fold-changes
# First let's label the genes as being significant in any condition or not.
res_shrunken_df <- res_shrunken_df %>%
  group_by(gene_id) %>%
  mutate(sig = ifelse(any(padj < 0.05), "sig", "ns"))

# Let's clean up the column names a little bit
res_shrunken_df <- res_shrunken_df %>%
  mutate(subcellular_fraction = gsub("condition_|_fraction_vs_total", "", result_name))

sig_res_shrunked_df <- res_shrunken_df %>%
  filter(sig == "sig")

lfc_matrix <- sig_res_shrunked_df %>% 
  dplyr::select(gene_id, log2FoldChange, subcellular_fraction) %>% 
  pivot_wider(names_from = "subcellular_fraction", values_from = "log2FoldChange") %>%
  column_to_rownames("gene_id") %>%
  as.matrix()

pheatmap::pheatmap(lfc_matrix, show_rownames = FALSE, breaks = seq(-3, 3, length.out = 100))
```

```{r}
#Generate each individual data.frame for each condition
#membrane_fraction_vs_total
res <- results(dds, name = "condition_membrane_fraction_vs_total")

res_shrunken_membrane <- lfcShrink(dds, coef = "condition_membrane_fraction_vs_total",  res = res)

res_shrunken_df_mem <- res_shrunken_membrane %>% as.data.frame() %>%
  rownames_to_column("gene_id") %>%
  merge(g2s) %>%
  mutate(result_name = "condition_membrane_fraction_vs_total")



#cytosolic_fraction_vs_total
res <- results(dds, name = "condition_cytosolic_fraction_vs_total")

res_shrunken_cytosolic <- lfcShrink(dds, coef = "condition_cytosolic_fraction_vs_total",  res = res)

res_shrunken_df_cyt <- res_shrunken_cytosolic %>% as.data.frame() %>%
  rownames_to_column("gene_id") %>%
  merge(g2s) %>%
  mutate(result_name = "condition_cytosolic_fraction_vs_total")


#insoluble_cytoplasmic_fraction_vs_total
res <- results(dds, name = "condition_insoluble_cytoplasmic_fraction_vs_total")

res_shrunken_insoluble_cytoplasmic <- lfcShrink(dds, coef = "condition_insoluble_cytoplasmic_fraction_vs_total",  res = res)

res_shrunken_df_ins <- res_shrunken_insoluble_cytoplasmic %>% as.data.frame() %>%
  rownames_to_column("gene_id") %>%
  merge(g2s) %>%
  mutate(result_name = "condition_insoluble_cytoplasmic_fraction_vs_total")




#nuclear_fraction_vs_total
res <- results(dds, name = "condition_nuclear_fraction_vs_total")

res_shrunken_nuclear <- lfcShrink(dds, coef = "condition_nuclear_fraction_vs_total",  res = res)

res_shrunken_df_nuc <- res_shrunken_nuclear %>% as.data.frame() %>%
  rownames_to_column("gene_id") %>%
  merge(g2s) %>%
  mutate(result_name = "condition_nuclear_fraction_vs_total")



```

```{r}
#merge all data.frames together
master_res_df <- rbind(res_shrunken_df_mem, res_shrunken_df_cyt, res_shrunken_df_ins, res_shrunken_df_nuc)

#pivot_wider specify values and names to make a big matrix
lfc_matrix <- master_res_df %>% dplyr::select(gene_id, log2FoldChange, result_name) %>%
  pivot_wider(names_from = "result_name", values_from = "log2FoldChange")

```

```{r}
#generate Heatmap
lfc_matrix_no_NA <- na.omit(lfc_matrix)
lfc_matrix_no_NA <- lfc_matrix_no_NA %>% column_to_rownames("gene_id") %>% as.matrix()

library(ComplexHeatmap)
pdf("figures/lfc_matrix_allconditions_heatmap.pdf", height = 35, width = 10)
Heatmap(lfc_matrix_no_NA,show_row_names = FALSE, show_column_names=TRUE)
dev.off()


