---
title: "Expression vs. DNA-binding at promoters"
output: html_document
editor_options:
  chunk_output_type: console
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
options(stringsAsFactors = F)
library(tidyverse)
library(GenomicRanges)
library(ggpubr)
source("util/intersect_functions.R")
source("util/_setup.R")
```


## Read in the sample sheet

```{r}
samplesheet <- read_csv("/scratch/Shares/rinnclass/JR/CLASS_2021/rnaseq/samplesheet.csv")
# The cleanest thing to do would be to fix it on the design file and re-run the nf-core pipeline,
# but for now we can swap it here.
samplesheet[which(samplesheet$sample_name == "hepg2_R1"), "condition"] <- "insoluble_cytoplasmic_fraction"
samplesheet[which(samplesheet$sample_name == "hepg2_insoluble_cytoplasmic_fraction_R2"), "condition"] <- "total"

#edit the path when you make changes
#write_csv("/scratch/Shares/rinnclass/data/samplesheet_fixed_condition.csv")
```

## Read in the TPM data from salmon

```{r}
salmon_tpm <- read_tsv("/scratch/Shares/rinnclass/JR/CLASS_2021/rnaseq/results/star_salmon/salmon.merged.gene_tpm.tsv")

# We want to summarize the TPM values for each subcellular fraction, so we'll take the 
# mean of the two replicates
tpm <- salmon_tpm %>% pivot_longer(cols = 2:ncol(.), names_to = "sample_name", values_to = "tpm") %>%
  merge(samplesheet) %>%
  group_by(gene_id, condition) %>%
  summarize(tpm = mean(tpm, na.rm = T)) %>%
  pivot_wider(names_from = condition, values_from = tpm, names_prefix = "tpm_")
```

## Read in the promoter features data

Which tells us how many DBPs are bound at each promoter

```{r}
promoter_features_df <- read.csv("/scratch/Shares/rinnclass/JR/CLASS_2021/analysis/01_global_peak_properties/results/peak_occurence_dataframe.csv")
# Now we can merge in the TPM data to this data.frame
promoter_features_df <- merge(promoter_features_df, tpm)

# Remember this plot!
ggplot(promoter_features_df, aes(x = number_of_dbp)) +
  geom_density()
```

## Plot a heatmap of where each gene is expressed

```{r}
# We need to turn this into a matrix
tpm_matrix <- tpm %>% 
  column_to_rownames("gene_id") %>%
  as.matrix()

# And z-scale each row.
tpm_scaled <- t(scale(t(tpm_matrix)))

# Remove NAs with the complete.cases function
?complete.cases
tpm_scaled <- tpm_scaled[complete.cases(tpm_scaled),]


pheatmap::pheatmap(tpm_scaled, show_rownames = FALSE)
pdf("figures/tpm_scaled_allconditions_heatmap.pdf", height = 40, width = 10)
```

## Plot the binding vs. expression

```{r}
# Total
# We're plotting only a trend line here with geom_smooth
# But we're also adding in the points for those with no expression (as triangle shapes)
# The ones with high number of DBPs and no expression are our reservoirs
pdf("figures/binding_vs_expression_plot.pdf", height = 10, width = 20)
ggplot(promoter_features_df, 
            aes(y = log2(tpm_total + 0.001), x = number_of_dbp, color = gene_type)) + 
  geom_point(data = promoter_features_df %>% filter(tpm_total < 0.001),
             shape = 17, alpha = 0.7) +
  geom_smooth(method = 'gam', formula = y ~ s(x, bs = "cs")) +
  # stat_cor() +
  scale_x_continuous(expand = c(0,0)) +
  scale_color_manual(values = c("#a8404c", "#424242"), name = "Gene type") + 
  ggtitle("Expression vs. promoter binding events") + 
  xlab(expression('Number of TFs')) +
  ylab(expression(log[2](TPM)))


```

```{r}
# Now plot this for each subcellular fraction
```


```{r}
# We can see this from the trend lines that lncRNAs are less expressed,
# but here's another way of looking at that.
ggplot(promoter_features_df, aes(x = log2(tpm_total + 0.01), color = gene_type))+
  geom_density()
```

## Now we can define reservoirs in HEPG2

```{r}
# We'll use a cutoff of 100 DBPs here, but we can think more about 
# what would be a reasonable cutoff here.
promoter_features_df$hepg2_reservoir <- 
  as.numeric(promoter_features_df$number_of_dbp > 100 & 
               promoter_features_df$tpm_total < 0.001)

table(promoter_features_df$hepg2_reservoir)
```

## Let's read in the reservoirs from last year

```{r}
k562_df <- read.csv("/scratch/Shares/rinnclass/data/2020_peak_occurence_df.csv")

# To compare these reservoirs to this year's dataset, we can just merge in the relevant columns
k562_reservoir <- k562_df %>% 
  dplyr::select(gene_id, reservoir, conservative_reservoir) %>%
  dplyr::rename(k562_reservoir = reservoir, 
                k562_conservative_reservoir = conservative_reservoir)

promoter_features_df <- merge(promoter_features_df, k562_reservoir)


# Make a table of reservoir status
res_status <- promoter_features_df %>% 
  group_by(hepg2_reservoir, k562_reservoir, k562_conservative_reservoir) %>%
  summarize(count = n())
```

```{r}
# We can now write these out for safekeeping / use in other analyses
write_csv(promoter_features_df, "results/promoter_features_df.csv")
write_csv(tpm, "results/mean_tpm_per_condition.csv")
write_csv(samplesheet, "results/samplesheet.csv")
```

There's a lot to explore here. We have some potential analyses that you can pursue, or your group can decide on a direction that you're interested in.

1. How does the binding vs. expression change for the high binding promoters (> ~300 DBPs) vs low binders? And does this trend change across subcellular fractions?
2. For the genes that are exclusively in one fraction (i.e. exclusively nuclear) what do their promoter binding profiles have in common?
3. Are there any mRNAs that are exclusively in the nuclear fraction?
4. What does the expression of the K562 reservoirs look like in HEPG2 in terms of expression? Are they more lowly expressed? Or does their trend follow the binding vs expression trend for non-K562 reservoirs?

```{r}
#just the hepg2_reservoirs and the k562_reservoirs
promoter_features_df$reservoirs_both <- 
  as.numeric(promoter_features_df$hepg2_reservoir == 1 & 
               promoter_features_df$k562_reservoir == 1)

#including the "conservative"
promoter_features_df$reservoirs_all <- 
  as.numeric(promoter_features_df$hepg2_reservoir == 1 & 
               promoter_features_df$k562_reservoir == 1 &
               promoter_features_df$k562_conservative_reservoir == 1)

```



```{r import}
source("/scratch/Shares/rinnclass/Chelsea/CLASS_2021/util/plotting_functions.R")
source("/scratch/Shares/rinnclass/Chelsea/CLASS_2021/util/intersect_functions.R")
source("/scratch/Shares/rinnclass/Chelsea/CLASS_2021/util/_setup.R")

gencode_gr <- rtracklayer::import("/scratch/Shares/rinnclass/data/gencode.v32.annotation.gtf")

fl <- list.files("/scratch/Shares/rinnclass/data/consensus_peaks", full.names = TRUE)
consensus_peaks <- lapply(fl, rtracklayer::import)
names(consensus_peaks) <- gsub("/scratch/Shares/rinnclass/data/consensus_peaks/|.bed", "", fl)
```

```{r make-promoters}
genes <- gencode_gr[gencode_gr$type == "gene"]
all_promoters_gr <- promoters(genes, upstream = 3e3, downstream = 3e3)
```

```{r}
# Let's look at what's binding on AL160408.1 and where it's binding.
AL160408.1_promoter <- all_promoters_gr[all_promoters_gr$gene_name == "AL160408.1"]
AL160408.1_promoter_matrix <- make_promoter_binding_matrix(consensus_peaks, AL160408.1_promoter)
plot_promoter_peak_matrix(AL160408.1_promoter_matrix, gene_name = "AL160408.1", save_pdf = TRUE)
```

