---
title: "Seperating out the mRNA and lncRNA Promoter binding events"
author: "CMT"
date: "03/12/2021"
output: html_document
editor_options: 
  chunk_output_type: console
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library(tidyverse)
library(GenomicRanges)
source("util/intersect_functions.R")
source("util/plotting_functions.R")
source("util/_setup.R")
library(Gviz)
library(ggpubr)
library(ggplot2)
```

```{r}
#Load in data for promoters
lncrna_mrna_promoters <- rtracklayer::import ("/scratch/Shares/rinnclass/Chelsea/analysis/results/promoter_regions/lncrna_mrna_promoters.gtf")

lncrna_promoters <- rtracklayer::import ("/scratch/Shares/rinnclass/Chelsea/analysis/results/promoter_regions/lncrna_promoters.gtf")

mrna_promoters <- rtracklayer::import ("/scratch/Shares/rinnclass/Chelsea/analysis/results/promoter_regions/mrna_promoters.gtf")
```


```{r}
#load in data for gene_bodies
lncrna_mrna_gene_bodies <- rtracklayer::import ("/scratch/Shares/rinnclass/Chelsea/analysis/results/gene_bodies/lncrna_mrna_genebody.gtf")

lncrna_promoters_gene_bodies <- rtracklayer::import ("/scratch/Shares/rinnclass/Chelsea/analysis/results/gene_bodies/lncrna_genebody.gtf")

mrna_promoters_gene_bodies <- rtracklayer::import ("/scratch/Shares/rinnclass/Chelsea/analysis/results/gene_bodies/mrna_genebody.gtf")
```


```{r}
#peak_occurence for both mRNA and lncRNA
peak_occurence <- read_csv("/scratch/Shares/rinnclass/Chelsea/analysis/results/peak_occurence_dataframe.csv")

#test <- read_table("/scratch/Shares/rinnclass/Chelsea/analysis/results/lncrna_mrna_promoter_peak_occurence_matrix.tsv")
```


```{r}
#Filter peak_occurence to separate mRNA and lnRNA
peak_occurence_lncRNA <- filter(peak_occurence, gene_type == "lncRNA")

#plot individual plot of peak_occurence_lnRNA
g <- ggplot(peak_occurence_lncRNA, aes(x = number_of_dbp))

g + geom_density(alpha = 0.2, color = "#424242", fill = "#424242") +
  theme_paperwhite() +
  xlab(expression("Number of DBPs")) +
  ylab(expression("Density")) +
  ggtitle("Promoter binding events",
          subtitle = "lncRNA genes") 

```

```{r}
#Filter peak_occurence to separate mRNA and lnRNA
peak_occurence_mRNA <- filter(peak_occurence, gene_type == "protein_coding")

#plot individual plot of peak_occurence_mRNA
g <- ggplot(peak_occurence_mRNA, aes(x = number_of_dbp))

g + geom_density(alpha = 0.2, color = "#424242", fill = "#424242") +
  theme_paperwhite() +
  xlab(expression("Number of DBPs")) +
  ylab(expression("Density")) +
  ggtitle("Promoter binding events",
          subtitle = "mRNA genes") 

```

```{r}
#plot both the lncRNA and protein_coding Promoter binding events
g <- ggplot(peak_occurence, aes(x = number_of_dbp, fill = gene_type, color = gene_type))

g + geom_density(alpha = 0.2, color = "#424242") +
theme_paperwhite() +
  xlab(expression("Number of DBPs")) +
  ylab(expression("Density")) +
  ggtitle("Promoter binding events",
          subtitle = "mRNA and lncRNA genes")
```

```{r}
#03.17.21 Determine what dbp are present in the two peaks (seen on peak occurance plot)
dbp_density_threshold_low <- 290
# Setting a threshold number in an object to be called below
#filtered_peak_occurence <- peak_occurence$number_of_dbp[peak_occurence$number_of_dbp > dbp_density_threshold]

second_peak_occurence <- filter(peak_occurence, peak_occurence$number_of_dbp > dbp_density_threshold_low)

dbp_density_threshold_high <- 310

middle_second_peak_occurence <- filter(second_peak_occurence, second_peak_occurence$number_of_dbp < dbp_density_threshold_high)

#Filter out the gene ID enriched with more than 200 bdp to paste into DAVID
bump2_promoters_geneID <- as.data.frame(middle_second_peak_occurence$gene_id)

write_csv(bump2_promoters_genenames, "/scratch/Shares/rinnclass/Chelsea/analysis/results/gene_id_second_peak_dbp.csv")

#Filter out the gene NAME 
bump2_promoters_genenames <- as.data.frame(middle_second_peak_occurence$gene_name)

write_csv(bump2_promoters_genenames, "/scratch/Shares/rinnclass/Chelsea/analysis/results/gene_name_second_peak_dbp.csv")
```

```{r}
#lncrna_mrna_promoter_peak_occurence <- read.table(file="lncrna_mrna_promoter_peak_occurence_matrix.tsv")

lncrna_mrna_promoter_peak_occurence_matrix <- read_table('/scratch/Shares/rinnclass/Chelsea/analysis/results/lncrna_mrna_promoter_peak_occurence_matrix.tsv')


#df_matrix <- as.data.frame(lncrna_mrna_promoter_peak_occurence)

#second_peak_promoters_matrix <- filter(df_matrix, colSums(df_matrix >200))

#Filter out the Zinc Finger proteins and generate a density plot
filtered_consensus_peaks_files <- list.files("/scratch/Shares/rinnclass/Chelsea/analysis/results/chipseq/filtered_consensus", 
                                             pattern = "*.bed",
                                             full.names = TRUE)
filtered_consensus_peaks <- lapply(filtered_consensus_peaks_files, rtracklayer::import)

names(filtered_consensus_peaks) <- gsub("/scratch/Shares/rinnclass/Chelsea/analysis/results/chipseq/filtered_consensus/|_filtered_consensus_peaks.bed", "", filtered_consensus_peaks_files)


#generate num_peaks_df
num_peaks_df <- data.frame("dbp" = names(filtered_consensus_peaks),
                           "num_peaks" = sapply(filtered_consensus_peaks, length))


#Filter out zinc fingers from the num_peaks_df data.frame
znf_num_peaks_df <- filter(num_peaks_df, num_peaks_df$dbd == "C2H2 ZF")



#Need to generate a peak_occurence matrix of the zinc finger proteins
##znf_promoter_peak_occurence <- count_peaks_per_feature(lncrna_mrna_promoters, filtered_consensus_peaks, type = "occurrence")


#filter the filtered_consensus_peaks using the znf_num_peaks_df
znf_filtered_consensus_peaks <- filter(filtered_consensus_peaks, num_peaks_df$dbp)


```

