---
title: "01_global_peak_properties"
author: "James Pratt"
date: "4/14/2021"
output: html_document
---
set working directory to the final analysis folder where all of the 0X folders are

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
options(stringsAsFactors = FALSE)
library(tidyverse)
library(GenomicRanges)
library(Gviz)
library(ggplot2)
source("../util/intersect_functions.R")
source("../util/plotting_functions.R")
source("../util/_setup.R")

```



converting tag matrix:
```{r}
#Load annotaitons
#ftp://ftp.ebi.ac.uk/pub/databases/gencode/Gencode_human/release_32/gencode.v32.annotation.gtf.gz
gencode_gr <- rtracklayer::import("data/gencode.v32.annotation.gtf")
g2s <- gencode_gr[gencode_gr$type == "gene"] %>%
  as.data.frame() %>%
  dplyr::select(gene_id, gene_name)

HEPG2_promoter_df <- read.table("results/lncrna_mrna_promoter_peak_occurence_matrix-1.tsv",
                              sep = " ", header = TRUE) %>%
  pivot_longer(2:ncol(.), names_to = "gene_id", values_to = "bound") %>%
  filter(bound == 1) %>%
  merge(g2s) %>%
  mutate(cell_type = "HEPG2")

k562_promoter_df <- read.table("results/k562_lncrna_mrna_promoter_peak_occurence_matrix-1.tsv",
                              sep = " ", header = TRUE) %>%
  pivot_longer(2:ncol(.), names_to = "gene_id", values_to = "bound") %>%
  filter(bound == 1) %>%
  merge(g2s) %>%
  mutate(cell_type = "K562")

promoter_df <- bind_rows(HEPG2_promoter_df, k562_promoter_df) %>%
  pivot_wider(names_from = cell_type, values_from = bound, values_fill = 0) %>%
  write_csv("promoter_dbp_map.csv")

firre_promoter <- promoter_df %>%
  filter(gene_name == "FIRRE", K562 == 1, HEPG2 == 1)

shf_promoter <- promoter_df %>%
  filter(gene_name == "SHF", K562 == 1, HEPG2 == 1)

shf_promoter$dbp[shf_promoter$dbp %in% firre_promoter$dbp]

```


global peak properties:
```{r}

filtered_consensus_peaks_files <- list.files("/scratch/Shares/rinnclass/data/filtered_consensus_peaks", 
                                             pattern = "*.bed",
                                             full.names = TRUE)
filtered_consensus_peaks <- lapply(filtered_consensus_peaks_files, rtracklayer::import)

names(filtered_consensus_peaks) <- gsub("/scratch/Shares/rinnclass/data/filtered_consensus_peaks/|_filtered_consensus_peaks.bed", "", filtered_consensus_peaks_files)

# Import annotations
lncrna_mrna_promoters <- rtracklayer::import("./00_consensus_peaks/lncrna_mrna_promoters.gtf")
lncrna_mrna_genebody <- rtracklayer::import("/scratch/Shares/rinnclass/YOUR_FOLDER/X/results/lncrna_mrna_genebody.gtf")

lncrna_gene_ids <- lncrna_mrna_genebody$gene_id[lncrna_mrna_genebody$gene_type == "lncRNA"]
mrna_gene_ids <- lncrna_mrna_genebody$gene_id[lncrna_mrna_genebody$gene_type == "protein_coding"]
```



```{r}
num_peaks_df <- data.frame("dbp" = names(filtered_consensus_peaks),
                           "num_peaks" = sapply(filtered_consensus_peaks, length))



num_peaks_df$total_peak_length <- sapply(filtered_consensus_peaks, function(x) sum(width(x)))

```

```{r}

promoter_peak_counts <- count_peaks_per_feature(lncrna_mrna_promoters, filtered_consensus_peaks, type = "counts")

num_peaks_df$peaks_overlapping_promoters <- rowSums(promoter_peak_counts)

num_peaks_df$peaks_overlapping_lncrna_promoters <- rowSums(promoter_peak_counts[,lncrna_gene_ids])

num_peaks_df$peaks_overlapping_mrna_promoters <- rowSums(promoter_peak_counts[,mrna_gene_ids])



```

```{r}



genebody_peak_counts <- count_peaks_per_feature(lncrna_mrna_genebody, 
                                                filtered_consensus_peaks, 
                                                type = "counts")

num_peaks_df$peaks_overlapping_genebody <- rowSums(genebody_peak_counts)
num_peaks_df$peaks_overlapping_lncrna_genebody <- rowSums(genebody_peak_counts[,lncrna_gene_ids])
num_peaks_df$peaks_overlapping_mrna_genebody <- rowSums(genebody_peak_counts[,mrna_gene_ids])


```


```{r}
# The human TFs
# https://www.cell.com/cms/10.1016/j.cell.2018.01.029/attachment/ede37821-fd6f-41b7-9a0e-9d5410855ae6/mmc2.xlsx

human_tfs <- readxl::read_excel("/scratch/Shares/rinnclass/data/mmc2.xlsx",
                                sheet = 2, skip = 1)
names(human_tfs)[4] <- "is_tf"

length(which(tolower(num_peaks_df$dbp) %in% tolower(human_tfs$Name)))


human_tfs <- human_tfs[tolower(human_tfs$Name) %in% tolower(num_peaks_df$dbp), 1:4]
names(human_tfs) <- c("ensembl_id",
                      "dbp",
                      "dbd",
                      "tf")

num_peaks_df <- merge(num_peaks_df, human_tfs, all.x = T)

# Let's check how many NAs -- we should have some missing values.

write_csv(num_peaks_df, "results/num_peaks_df.csv")
```


```{r}
promoter_peak_occurence <- count_peaks_per_feature(lncrna_mrna_promoters, filtered_consensus_peaks, 
                                               type = "occurrence")
# Output to promoter_peak_occurecne_matrix
write.table(promoter_peak_occurence, "results/lncrna_mrna_promoter_peak_occurence_matrix.tsv")
# Now we want to make into a data frame using the promoter annotations as rows and attributes as columns.
# We will use lncrna_mrna_promoters to index "all promoters"
# First make sure promoter_peak_occurence and lncrna_mrna_promoters are in the same order
stopifnot(all(colnames(promoter_peak_occurence) == lncrna_mrna_promoters$gene_id))



# We are going to use the promoter peak occurence matrix above to essentially
# recreate a working version of num_peaks_df. However we will now organize it more
# this is a good example of how to set up a bunch of columns in dataframe. using the
# data.frame() fucntion.
# essentially we will index values from the objects we created and make a .CSV 
# to keep adding onto in future classes.

peak_occurence_df <- data.frame("gene_id" = colnames(promoter_peak_occurence),
                                "gene_name" = lncrna_mrna_promoters$gene_name,
                                "gene_type" = lncrna_mrna_promoters$gene_type,
                                "chr" = lncrna_mrna_promoters@seqnames,   
                                "3kb_up_tss_start" = lncrna_mrna_promoters@ranges@start,
                                "strand" = lncrna_mrna_promoters@strand,
                                "number_of_dbp" = colSums(promoter_peak_occurence))
# This is the CSV file we will start building upon adding columns of properties as we analyze them
# The output file name will change based on what is added later, but the "peak_occurence_df" will be used throughout.
write_csv(peak_occurence_df, "results/peak_occurence_dataframe.csv")
```


class fix:
```{r}

occurence_matrix <- read.table("results/lncrna_mrna_promoter_peak_occurence_matrix.tsv")

super_promoter <- occurence_matrix[, colSums(occurence_matrix) > 350]

super_promoter_dbps <- data.frame(dbp = rownames(super_promoter), num_super_promoter_overlaps = rowSums(super_promoter))

num_peaks_df <- read.csv("results/num_peaks_df.csv")

super_prom_features <- merge(super_promoter_dbps, num_peaks_df)

super_promoter_features <- dplyr::select(super_promoter_features, dbp, num_super_promoter_overlaps, everything() )

super_promoter_features <- super_promoter_features %>%
  mutate(percent_super_prom_ov = num_super_promoter_overlaps/peaks_overlapping_promoters * 100, 
         percent_of_total_peaks_at_super_prom = num_super_promoter_overlaps/num_peaks * 100,
         zincfinger = ifelse(grepl(" ZF",dbd), "ZF", "Other"))

table(super_promoter_features$zincfinger)

ggplot(super_promoter_features, aes(x = percent_super_prom_ov, color = zincfinger)) +
  geom_density()
ggplot(super_promoter_features, aes(x = percent_of_total_peaks_at_super_prom, y = percent_super_prom_ov)) + 
  geom_point()

write.csv(super_promoter_features, "results/super_promoter_features.csv")
  
```
