---
title: "14-2: RNA-seq differential expression"
output: html_document
editor_options: 
  chunk_output_type: console
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
options(stringsAsFactors = FALSE)
library(tidyverse)
library(tximport)
library(DESeq2)
```

Read in the count data.

```{r}
hepg2_counts <- read_tsv("data/GeneCountMatrix.tsv")
# Truncate the gene id to match the format from dee2
gene_id_trunc <- sapply(genes$gene_id, function(x) unlist(strsplit(x, ".", fixed = T))[[1]])
gene_id_map <- data.frame(gene_id = genes$gene_id,
                          GeneID = gene_id_trunc)
table(hepg2_counts$GeneID %in% gene_id_map$GeneID)
hepg2_counts <- hepg2_counts %>%
  filter(GeneID %in% gene_id_map$GeneID) %>%
  merge(gene_id_map) %>%
  dplyr::select(-GeneID)
# Merge this with our data from ENCODE
files <- file.path("../../../rinn/JR/CLASS_2021/rnaseq/results/star_salmon",
                   samplesheet$sample_name, 
                   "quant.sf")
names(files) <- samplesheet$sample_name
txi <- tximport(files, type = c("salmon"), tx2gene = tx2gene)
encode_counts <- txi$counts %>%
  as.data.frame() %>%
  rownames_to_column("gene_id")
counts <- merge(encode_counts, hepg2_counts)
counts <- counts %>% 
  column_to_rownames("gene_id") %>%
  as.matrix()
# DEseq2 wants an integer count matrix
mode(counts) <- "integer"
# Need to make a samplesheet for this data.
hep2g_meta <- read_tsv("data/MetadataSummary.tsv") %>%
  dplyr::select(SRR_accession, Library_name) %>%
  dplyr::rename(sample_name = SRR_accession, 
                sample_title = Library_name) %>%
  mutate(condition = "DEE2_HEPG2")
samples <- samplesheet %>%
  mutate(sample_title = sample_name) %>%
  dplyr::select(sample_name, sample_title, condition)
samples <- samples %>%
  bind_rows(hep2g_meta)
# Make sure the sample dataset is in the proper order
rownames(samples) <- samples$sample_name
samples <- samples[colnames(counts),]
stopifnot(all(rownames(samples) == colnames(count)))
samples$condition <- factor(samples$condition, levels = c("total","nuclear_fraction","cytosolic_fraction",
                                                            "insoluble_cytoplasmic_fraction", 
                                                            "membrane_fraction",  "DEE2_HEPG2"))
# Change the count data to integer format
dds <- DESeqDataSetFromMatrix(countData = counts,
                              colData = samples,
                              design = ~condition)
dds <- DESeq(dds)
rlog_counts <- rlog(dds)
rld_pca <- prcomp(t(assay(rlog_counts)))
rld_prcomps <- rld_pca$x %>% as.data.frame() %>%
  rownames_to_column("sample_name") %>%
  select(sample_name, PC1, PC2)
rld_prcomps <- merge(samples, rld_prcomps)
g <- ggplot(rld_prcomps, aes(x = PC1, y = PC2, color = condition))
g + geom_point() 
```


```{r}
hepg2_counts$GeneID %in% genes$gene_id
gencode_gtf <- rtracklayer::import("/scratch/Shares/rinn/genomes/Homo_sapiens/Gencode/v32/gencode.v32.annotation.gtf")
genes <- gencode_gtf[gencode_gtf$type == "gene"]
tx2gene <- gencode_gtf[gencode_gtf$type == "transcript"]@elementMetadata %>% 
  as.data.frame() %>%
  dplyr::select(transcript_id, gene_id)
samplesheet <- read_csv("../../../rinn/JR/CLASS_2021/rnaseq/samplesheet.csv")
rownames(samplesheet) <- samplesheet$sample_name
hmm <- txi$abundance
samplesheet <- samplesheet %>%
  mutate(condition = factor(condition, levels = c("total",
                                                  "nuclear_fraction",
                                                  "cytosolic_fraction",
                                                  "membrane_fraction",
                                                  "insoluble_cytoplasmic_fraction")))
# Reorder the rows and columns to match
names(genes) <- genes$gene_id
genes <- genes[rownames(txi$counts)]
stopifnot(all(genes$gene_id == rownames(txi$counts)))
samplesheet <- samplesheet %>%
  mutate(row_names = sample_name) %>%
  column_to_rownames("row_names")
samplesheet <- samplesheet[colnames(txi$counts),]
stopifnot(all(rownames(samplesheet) == colnames(txi$counts)))
dds <- DESeqDataSetFromTximport(txi,
                                colData = samplesheet,
                                design = ~ condition,
                                rowData = genes)
dds <- DESeq(dds)
```

```{r}
rlog_counts <- rlog(dds)
```

```{r}
resultsNames(dds)
res <- results(dds, name = "condition_cytosolic_fraction_vs_total")
rld_pca <- prcomp(t(assay(rlog_counts)))
rld_prcomps <- rld_pca$x %>% as.data.frame() %>%
  rownames_to_column("sample_name") %>%
  select(sample_name, PC1, PC2, PC3)
rld_prcomps <- merge(samplesheet, rld_prcomps)
g <- ggplot(rld_prcomps, aes(x = PC1, y = PC2, color = condition))
g + geom_point() 
```