---
title: "Expression vs. DNA-binding at promoters"
output: html_document
editor_options:
  chunk_output_type: console
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
options(stringsAsFactors = F)
library(tidyverse)
library(GenomicRanges)
library(GenomicFeatures)
library(rtracklayer)
library(ggrepel)
library(ggpubr)
library(Rsubread)
library(regioneR)
source("../../util/intersect_functions.R")
source("../../util/_setup.R")
```


Reading in peak occurence DF and Res_df
```{r}

promoter_features_df <- read.csv("/scratch/Shares/rinnclass/JR/CLASS_2021/analysis/01_global_peak_properties/results/peak_occurence_dataframe.csv")

samplesheet <- read_csv("/scratch/Shares/rinnclass/JR/CLASS_2021/rnaseq/samplesheet.csv")

# We of course need to swap the labels for total and insoluble cytoplasmic
# From looking in the browser at genes that should be nuclear: NEAT1 for ex.
# We have concluded that 
# hepg2_R2 -- is whole cell / total
# hepg2_insoluble_cytoplasmic_fraction_R2 -- is whole cell / total
# hepg2_R1 -- is insoluble_cytoplasmic
# hepg2_insoluble_cytoplasmic_fraction_R1 -- is insoluble_cytoplasmic

# The cleanest thing to do would be to fix it on the design file and re-run the nf-core pipeline,
# but for now we can swap it here.
samplesheet[which(samplesheet$sample_name == "hepg2_R1"), "condition"] <- "insoluble_cytoplasmic_fraction"
samplesheet[which(samplesheet$sample_name == "hepg2_insoluble_cytoplasmic_fraction_R2"), "condition"] <- "total"


RNA_seq_data <- read_tsv("/scratch/Shares/rinnclass/JR/CLASS_2021/rnaseq/results/star_salmon/salmon.merged.gene_tpm.tsv")

tpm <- RNA_seq_data %>% pivot_longer(cols = 2:ncol(.), names_to = "sample_name", values_to = "tpm") %>%
  merge(samplesheet) %>%
  group_by(gene_id, condition) %>%
  summarize(tpm = mean(tpm, na.rm = T)) %>%
  pivot_wider(names_from = condition, values_from = tpm, names_prefix = "tpm_")

promoter_features_df <- merge(promoter_features_df, tpm)

```
Writing out useful files from above
```{r}

write_csv(tpm, "results/mean_tpm_per_condition.csv")

write_csv(samplesheet, "results/sample_sheet.csv")


```

Total
```{r}
g <- ggplot(promoter_features_df, 
            aes(y = log2(tpm_total + 0.001), x = number_of_dbp, color = gene_type))
g + geom_point(data = promoter_features_df %>% filter(tpm_total < 0.001),
             shape = 17, alpha = 0.7) +
  geom_smooth(method = 'gam', formula = y ~ s(x, bs = "cs")) +
  stat_cor() +
  scale_x_continuous(expand = c(0,0)) +
  scale_color_manual(values = c("#a8404c", "#424242"), name = "Gene type") + 
  ggtitle("Expression vs. promoter binding events") + 
  xlab(expression('Number of TFs')) +
  ylab(expression(log[2](TPM)))
```

Nuclear
```{r}
g <- ggplot(promoter_features_df, 
            aes(y = log2(tpm_nuclear_fraction + 0.001), x = number_of_dbp, color = gene_type))
g + geom_point(data = promoter_features_df %>% filter(tpm_nuclear_fraction < 0.001),
             shape = 17, alpha = 0.7) +
  geom_smooth(method = 'gam', formula = y ~ s(x, bs = "cs")) +
  stat_cor() +
  scale_x_continuous(expand = c(0,0)) +
  scale_color_manual(values = c("#a8404c", "#424242"), name = "Gene type") + 
  ggtitle("Expression vs. promoter binding events") + 
  xlab(expression('Number of TFs')) +
  ylab(expression(log[2](TPM)))
```

Membrane
```{r}
g <- ggplot(promoter_features_df, 
            aes(y = log2(tpm_membrane_fraction + 0.001), x = number_of_dbp, color = gene_type))
g + geom_point(data = promoter_features_df %>% filter(tpm_membrane_fraction < 0.001),
             shape = 17, alpha = 0.7) +
  geom_smooth(method = 'gam', formula = y ~ s(x, bs = "cs")) +
  stat_cor() +
  scale_x_continuous(expand = c(0,0)) +
  scale_color_manual(values = c("#a8404c", "#424242"), name = "Gene type") + 
  ggtitle("Expression vs. promoter binding events") + 
  xlab(expression('Number of TFs')) +
  ylab(expression(log[2](TPM)))
```

Cytodsolic
```{r}
g <- ggplot(promoter_features_df, 
            aes(y = log2(tpm_cytosolic_fraction + 0.001), x = number_of_dbp, color = gene_type))
g + geom_point(data = promoter_features_df %>% filter(tpm_cytosolic_fraction < 0.001),
             shape = 17, alpha = 0.7) +
  geom_smooth(method = 'gam', formula = y ~ s(x, bs = "cs")) +
  stat_cor() +
  scale_x_continuous(expand = c(0,0)) +
  scale_color_manual(values = c("#a8404c", "#424242"), name = "Gene type") + 
  ggtitle("Expression vs. promoter binding events") + 
  xlab(expression('Number of TFs')) +
  ylab(expression(log[2](TPM)))
```

Insoluble
```{r}
g <- ggplot(promoter_features_df, 
            aes(y = log2(tpm_insoluble_cytoplasmic_fraction + 0.001), x = number_of_dbp, color = gene_type))
g + geom_point(data = promoter_features_df %>% filter(tpm_insoluble_cytoplasmic_fraction < 0.001),
             shape = 17, alpha = 0.7) +
  geom_smooth(method = 'gam', formula = y ~ s(x, bs = "cs")) +
  stat_cor() +
  scale_x_continuous(expand = c(0,0)) +
  scale_color_manual(values = c("#a8404c", "#424242"), name = "Gene type") + 
  ggtitle("Expression vs. promoter binding events") + 
  xlab(expression('Number of TFs')) +
  ylab(expression(log[2](TPM)))
```


```{r}
ggplot(promoter_features_df, aes(x = log2(tpm_total + 0.01), color = gene_type))+
  geom_density()
```

```{r}
promoter_features_df$reservoir <- 
  as.numeric(promoter_features_df$number_of_dbp > 100 & 
               promoter_features_df$tpm_total < 0.001)

table(promoter_features_df$reservoir)

promoter_features_df %>%
  mutate(overlapping_reservoir = promoter_features_df$reservoir %in% k562_df$reservoir)

table(promoter_features_df$overlapping_reservoir)


```


Reading in reservoirs from last year
```{r}
k562_df <- read.csv("/scratch/Shares/rinnclass/data/2020_peak_occurence_df.csv")
k562_df <- read.csv("/Shares/rinn_class/data/CLASS_2020/analysis/08_defining_reservoirs/results/08_peak_occurence_df_resvoirs.csv")
k562_reservoir <- k562_df %>% dplyr::select(gene_id, reservoir, conservative_reservoir) %>%
  dplyr::rename(k562_reservoir = reservoir, k562_conservative_reservoir = conservative_reservoir)

promoter_features_df <- read.csv("results/promoter_feature_df.csv")
promoter_features_df <- merge(promoter_features_df, k562_reservoir)


# Make a table of reservoir status
res_status <- promoter_features_df %>% 
  group_by(reservoir, k562_reservoir, k562_conservative_reservoir) %>%
  summarize(count = n())
  

table(promoter_features_df$k562_reservoir)

write_csv(promoter_features_df, "results/promoter_features_df.csv")
```

