---
title: "Global peak properties"
output: html_document
editor_options: 
  chunk_output_type: console
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library(tidyverse)
library(GenomicRanges)
library(Gviz)
source("/scratch/Shares/rinnclass/tardigrades/CLASS_2021/util/intersect_functions.R")
source("/scratch/Shares/rinnclass/tardigrades/CLASS_2021/util/_setup.R")
```

# Load Filtered Consensus Peaks, Promoter Regions, & Gene Body to make genebody occurance and genebody-promoter occurance


```{r}
# Import filtered_consensus_peaks_files t omake new occurances
filtered_consensus_peaks_files <- list.files("/scratch/Shares/rinnclass/tardigrades/CLASS_2021/analysis/00_consensus_peaks/results/filtered_consensus", 
                                             pattern = "*.bed",
                                             full.names = TRUE)
filtered_consensus_peaks <- lapply(filtered_consensus_peaks_files, rtracklayer::import)

names(filtered_consensus_peaks) <- gsub("/scratch/Shares/rinnclass/tardigrades/CLASS_2021/analysis/00_consensus_peaks/results/filtered_consensus/|_filtered_consensus_peaks.bed", "", filtered_consensus_peaks_files)

# Import genebody annotations
lncrna_mrna_genebody <- rtracklayer::import("/scratch/Shares/rinnclass/tardigrades/CLASS_2021/analysis/00_consensus_peaks/results/genebody/lncrna_mrna_genebody.gtf")

gencode_gr <- rtracklayer::import("/Shares/rinn_class/data/genomes/human/gencode/v32/gencode.v32.annotation.gtf")
```

# Generate Genebody Peak Occurence like we did in 01 for promoters

```{r}
genebody_peak_occurence <- count_peaks_per_feature(lncrna_mrna_genebody, filtered_consensus_peaks, 
                                               type = "occurrence")

# Output to genebody_peak_occurence matrix
write.table(genebody_peak_occurence, "/scratch/Shares/rinnclass/tardigrades/CLASS_2021/analysis/01_global_peak_properties/results/lncrna_mrna_genebody_peak_occurence_matrix.tsv")

# First make sure genebody_peak_occurence and lncrna_mrna_genebody are in the same order
stopifnot(all(colnames(genebody_peak_occurence) == lncrna_mrna_genebody$gene_id))


# We are going to use the promoter peak occurence matrix above to keep adding onto in future files.
genebody_peak_occurence_df <- data.frame("gene_id" = colnames(genebody_peak_occurence),
                                "gene_name" = lncrna_mrna_genebody$gene_name,
                                "gene_type" = lncrna_mrna_genebody$gene_type,
                                "chr" = lncrna_mrna_genebody@seqnames,   
                                "3kb_up_tss_start" = lncrna_mrna_genebody@ranges@start,
                                "gene_length" = lncrna_mrna_genebody@ranges@width,
                                "strand" = lncrna_mrna_genebody@strand,
                                "number_of_dbp" = colSums(genebody_peak_occurence))

write_csv(genebody_peak_occurence_df, "/scratch/Shares/rinnclass/tardigrades/CLASS_2021/analysis/01_global_peak_properties/results/genebody_peak_occurence_dataframe.csv")
```

# Generate Genebody minus promoter Peak Occurence df

```{r}
# PLUS strand
# filter out all genes that are <3000 bp
plus_strand_genes <- gencode_gr[strand(gencode_gr) == "+" & gencode_gr$type == "gene" & 
                                     gencode_gr$gene_type %in% c("lncRNA", "protein_coding")&
                                gencode_gr@ranges@width > 3000]
# define new gene body region starting at +3000 untill end of gene; calculate new gene lenght minus 3000
gene_body_plus <- GRanges(seqnames = seqnames(plus_strand_genes), ranges = IRanges(start = start(plus_strand_genes) + 3000, end = end(plus_strand_genes), width = width(plus_strand_genes) - 3000))
# put back into matrix to merge mack to matrix with metadata
plus_strand_genes@ranges <- gene_body_plus@ranges

# same for minus strand
min_strand_genes <- gencode_gr[strand(gencode_gr) == "-" & gencode_gr$type == "gene" & 
                                     gencode_gr$gene_type %in% c("lncRNA", "protein_coding")&
                                gencode_gr@ranges@width > 3000]

gene_body_minus <- GRanges(seqnames = seqnames(min_strand_genes), ranges = IRanges(start = start(min_strand_genes) + 3000, end = end(min_strand_genes), width = width(min_strand_genes) - 3000))

min_strand_genes@ranges <- gene_body_minus@ranges

# calculate the occurances for both strands
plus_occ <- count_peaks_per_feature(plus_strand_genes, filtered_consensus_peaks, 
                                               type = "occurrence")
#  make sure plus_occ and plus_strand_genes are in the same order
stopifnot(all(colnames(plus_occ) == plus_strand_genes$gene_id))

min_occ <- count_peaks_per_feature(min_strand_genes, filtered_consensus_peaks, 
                                               type = "occurrence")
#  make sure plus_occ and plus_strand_genes are in the same order
stopifnot(all(colnames(min_occ) == min_strand_genes$gene_id))

# make df
plus_df <- data.frame("gene_id" = colnames(plus_occ),
                                "gene_name" = plus_strand_genes$gene_name,
                                "gene_type" = plus_strand_genes$gene_type,
                                "chr" = plus_strand_genes@seqnames,   
                                "3kb_up_tss_start" = plus_strand_genes@ranges@start,
                      "new_gene_length" = plus_strand_genes@ranges@width,
                                "strand" = plus_strand_genes@strand,
                                "number_of_dbp" = colSums(plus_occ))
min_df <- data.frame("gene_id" = colnames(min_occ),
                                "gene_name" = min_strand_genes$gene_name,
                                "gene_type" = min_strand_genes$gene_type,
                                "chr" = min_strand_genes@seqnames,   
                                "3kb_up_tss_start" = min_strand_genes@ranges@start,
                     "new_gene_length" = min_strand_genes@ranges@width,
                                "strand" = min_strand_genes@strand,
                                "number_of_dbp" = colSums(min_occ))
#combine both df
genebodyMinusPromoter_df <- rbind(plus_df,min_df)
# output df
write_csv(genebodyMinusPromoter_df, "/scratch/Shares/rinnclass/tardigrades/CLASS_2021/analysis/01_global_peak_properties/results/GgenebodyMinusPromoter_df.csv")

# or combine both peak occurances and save them
genebodyMinuspromoter_peak_occurence <- cbind(plus_occ,min_occ)

# Output to genebodyMinuspromoter_peak_occurence_matrix
write.table(genebodyMinuspromoter_peak_occurence, "/scratch/Shares/rinnclass/tardigrades/CLASS_2021/analysis/01_global_peak_properties/results/GenebodyMinuspromoter_peak_occurence_matrix.tsv")


```

