---
title: "00_global_clustering"
author: "Group 2"
date: "4/19/2021"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library(tidyverse)
library(ggdendro)
library(GenomicRanges)
source("../../util/intersect_functions.R")
source("../../util/plotting_functions.R")
source("../../util/_setup.R")
```

The goal for today is to "cluster" the 438 DBPs to find which ones have similar binding profiles across the genome -- and those that are different. 

Let's start by refreshing our memory of what all went down in 01_global_peak_properties.

Briefly we did 2 really good things that we will need:

1) We made a list of (i) all promoters (ii) lncRNA promoters only (iii) mRNA promoters
2) We made a peak_occurance_dataframe (.tsv) that is comprised of columns for each dbp. Rows are the DBP and filled 1 or 0 if that DBP binds to a given promoter. Or:

Promoter 1, 2, 3, ....
DBP      0, 0, 1,....

Thus, each row has ~36,000 1/0 for a given DBP. This is binary data (1/0) and not contious data. We need to consider that when clustering as these similarity measurements will be influenced by continious or binary data.

We exported this file to "results" folder and same for the GTFs. So we are all set to start clustering.



There are numerous ways to do clustering, but the basic principle is that the algorithm will start from the bottom and move up. This means it will find to samples with highest similarity and then find the next and the next.. in agglormative manner.

Really all this is a correlation (e.g. Pearson) between two vectors of ~60,000 values of 1/0. 

Let's get started by clustering all promoters first, then later we can see how this changes when we seperate lncRNAs and mRNAs.

```{r Clustering all Promoters}
# Let's grab both the lncRNA and mRNA promoters
# promoters <- rtracklayer::import("results/lncrna_mrna_promoters.gtf")
peak_occurence_matrix <- read.table("../01_global_peak_properties/results/lncrna_mrna_promoter_peak_occurence_matrix.tsv")

# Let's filter again to make sure we have at least 250 peaks
# This is a different filtering because we're now filtering on at least 250 peaks on 
# promoter regions rather than 250 peaks across the genome.
peak_occurence_matrix <- peak_occurence_matrix[rowSums(peak_occurence_matrix) > 250, ]

# ok now we have all the things we need to start clustering !
# First we will use hclust

# ?hclust
# bin_hier
# R is so savy with statistics that this is actually a base R function!
# There are lots of methods to cluster by let's start with binary.

# Clustering requires a distance matrix which represents each row in the matrix.
# In this case rows are DBPs. We could also cluster on the columns (promoters) but
# for now we'll start with the DBPs.
# If we just want to calculate the euclidean distance between the first two rows,
# we can use the dist function.
peak_occurence_matrix[1:2,1:10]
dist(peak_occurence_matrix[1:2,], method = "euclidean")


# Since this is a binary vector, let's use the binary distance measure.
# The binary measure only considers those elements which are not zero in both vectors.
# It is then the proportion of elements which are not shared (i.e. 0 in vector 1, and 1 in vector 2).
dist(peak_occurence_matrix[1:4,], method = "binary")


# We can create a distance matrix for each pairwise vector comparison for the whole matrix now.
# This will result in a 460x460 matrix that is symmetric on either side of the diagonal.
peak_occurence_dist <- dist(peak_occurence_matrix, method = "binary")

# Now that we have that, we can do hierarchical clustering. In brief, this will cluster more closely
# vectors that have small distances, and will iterate that process until everything is in the same cluster.
# Hierarchical clustering with binary distance measure
bin_hier <- hclust(peak_occurence_dist, method = "complete")

plot(bin_hier)

# To make this bigger, we can plot it as a pdf
pdf("figures/dbp_hclust_dendro.pdf", height = 12, width = 70)
plot(bin_hier)
dev.off()

bin_hier$labels[bin_hier$order]
```

Nice so we now have object "bin_hier" that will contain all the cross correlation values across all the samples. 

Now we get to PLOT !! We will use ggdenro package that will plot the branch lengths that indicate how similar two samples are -- the longer the branch the more different they are -- closer more similar.

Here is an alternative way to plot it:

```{r Plot CLuster using ggdendro}


# First we are going to order the samples by the correlation matrix we just made.

# You can also use ggdendro to plot with ggplot syntax -- which looks nice but for this many 
# DBPs, it's a little bit easier to visualize with base R.

# Now we can plot
ggdendro::ggdendrogram(bin_hier, rotate = FALSE,  size = 3, 
                        theme_dendro = TRUE) +
   coord_flip() +
    scale_y_continuous() +
    scale_x_continuous(position = "top") +
    scale_x_continuous(breaks = seq_along(bin_hier$labels[bin_hier$order]),
             labels = bin_hier$labels[bin_hier$order], position = "top",
             expand = c(0,0)) +
   theme(axis.text.x = element_text(angle = 90, hjust  = 1)) + 
   theme(axis.text.y = element_text(angle = 0,hjust = 1)) +
   scale_y_reverse(expand = c(0.01, 0)) +
   theme(
     plot.background = element_blank(),
     panel.grid.major = element_blank(),
     panel.grid.minor = element_blank(),
     panel.border = element_blank()
   )

# Go over the plotting aspects as we discussed in 11_intro_plotting_R.Rmd
# Especially thje scale_x_continious part -- we saw that in the 11_ class 

# So this is very difficult to see since we have SO much data :)
# one thing you can see is "phospho S5" and "phospho S2".
# These are modifications on Pol II that indicate two different states:
# S2 = distal or elongating Pol II so less on promoter more in genebody
# S5 = Indicative of Pol II at promoter proximilar regions.
# Note how these don't cluster anywhere near each other -- that is good news!


# Let's save this as a figure -- we want to keep organized so we made a ChIPseq
# folder and figures inside that folder. Let's put the figure there:

ggsave("figures/all_hclust_binary_dist.pdf", height = 44, width = 5.0)

# Nice, this will make it much easier to read :)
# use file transfer to bring to desktop and take a look!
# This is looking biologically relevant! Notice Pol II phospho 2 is
# similar to all other compents of Pol II -- that is exactly what we expect.
# also Pol II S5 is associated with H3K27ac which is a mark of active promoters
# again what we would expect of promoter proximal Pol II.
# MED DBPs all near each other too!
# Sweet -- with a simple spot check we can see all of this data is making sense!
```

Here's some code to cluster mRNA promoters and lncRNA promoters separately. They end up looking pretty similar.

```{r Cluster mRNA promoters and lncRNA promoters separately}

# # Let's start with lncRNA promoters -- why not?
# 
# # We first need to load the .GTF we made for lncRNAs and mRNAs.
# 
lncrna_promoters <- rtracklayer::import("../00_consensus_peaks/results/lncrna_promoters.gtf")
mrna_promoters <- rtracklayer::import("../00_consensus_peaks/results/mrna_promoters.gtf")
 
# Cool now we can do the same as above:
 
lncrna_peak_occurence <- peak_occurence_matrix[,lncrna_promoters$gene_id]
 
bin_hier_lncrna <- hclust(dist(lncrna_peak_occurence, method = "binary"))
 
ggdendro::ggdendrogram(bin_hier_lncrna, rotate = TRUE,  size = 3)
 
ggsave("figures/lncrna_hclust_binary_dist.pdf", height = 44, width = 6)
 
#Cool we can see the Pol II S2 and S5 diff right away again!
 
 
#Now for mRNA
 
mrna_peak_occurence <- peak_occurence_matrix[,mrna_promoters$gene_id]
 
bin_hier_mrna <- hclust(dist(lncrna_peak_occurence, method = "binary"))
 
ggdendro::ggdendrogram(bin_hier, rotate = TRUE,  size = 3)
 
ggsave("figures/mrna_hclust_binary_dist.pdf", height = 44, width = 6)


# Do a file transfer and open up lncRNA and mRNA clusters side by side:
# They are almost identical !! That is another good sign.

```


