---
title: "TSS profiles: plotting & clustering"
output: html_document
editor_options: 
  chunk_output_type: console
---

```{r setup, include=FALSE}
options(stringsAsFactors = FALSE)
knitr::opts_chunk$set(echo = TRUE)
library(GenomicRanges)
library(rtracklayer)
library(tidyverse)

source("util/plotting_functions.R")
source("util/intersect_functions.R")
source("util/_setup.R")
```

# TSS profile plots

## Import the data we need

We'll be using the consensus peaks we generated previously (MACS peak calls filtered to those that are overlapping across replicates) and the Gencode gene annotations. 

```{r import}
gencode_gr <- rtracklayer::import("/scratch/Shares/rinnclass/data/gencode.v32.annotation.gtf")

fl <- list.files("/scratch/Shares/rinnclass/data/consensus_peaks", full.names = TRUE)
consensus_peaks <- lapply(fl, rtracklayer::import)
names(consensus_peaks) <- gsub("/scratch/Shares/rinnclass/data/consensus_peaks/|.bed", "", fl)
```

```{r}
genes <- gencode_gr[gencode_gr$type == "gene"]
# This function is a convenience function built into GenomicRanges
all_promoters_gr <- promoters(genes, upstream = 3e3, downstream = 3e3)
table(width(all_promoters_gr))
```

# Let's put all this into a funciton so we can call it later conveniently.

```{r}
profile_tss <- function(peaks, 
                        promoters_gr,
                        upstream = 3e3,
                        downstream = 3e3) {
  

  peak_coverage <- coverage(peaks)
  
  coverage_length <- elementNROWS(peak_coverage)
  coverage_gr <- GRanges(seqnames = names(coverage_length),
                         IRanges(start = rep(1, length(coverage_length)), 
                                 end = coverage_length))
  
  promoters_gr <- subsetByOverlaps(promoters_gr, 
                                       coverage_gr, 
                                       type="within", 
                                       ignore.strand=TRUE)
  chromosomes <- intersect(names(peak_coverage), 
                           unique(as.character(seqnames(promoters_gr))))
  peak_coverage <- peak_coverage[chromosomes]
  
  promoters_ir <- as(promoters_gr, "IntegerRangesList")[chromosomes]
  
  promoter_peak_view <- Views(peak_coverage, promoters_ir)
  
  promoter_peak_view <- lapply(promoter_peak_view, function(x) t(viewApply(x, as.vector)))
  promoter_peak_matrix <- do.call("rbind", promoter_peak_view)
  
  minus_idx <- which(as.character(strand(promoters_gr)) == "-")
  promoter_peak_matrix[minus_idx,] <- promoter_peak_matrix[minus_idx,
                                                           ncol(promoter_peak_matrix):1]
  
  promoter_peak_matrix <- promoter_peak_matrix[rowSums(promoter_peak_matrix) > 1,]
  
  peak_sums <- colSums(promoter_peak_matrix)
  peak_dens <- peak_sums/sum(peak_sums)
  
  metaplot_df <- data.frame(x = -upstream:(downstream-1),
                            dens = peak_dens)
  
  return(metaplot_df)
}
```

## Use this function to make separate plots for lncRNA and mRNA

First we'll create separate objects for lncRNA promoters and mRNA promoters, then we'll supply each of these to the new function we just made.

```{r}
# lncRNA promoter
lncrna_genes <- genes[genes$gene_type == "lncRNA"]
lncrna_promoters <- promoters(lncrna_genes, upstream = 3e3, downstream = 3e3)
# mRNA promoter
mrna_genes <- genes[genes$gene_type == "protein_coding"]
mrna_promoters <- promoters(mrna_genes, upstream = 3e3, downstream = 3e3)

for (i in 1:length(genes_of_interest)){
  current_gene <- genes_of_interest[i]
  all_genes_metaplot_profile <- profile_tss(consensus_peaks[[current_gene]], all_promoters_gr)
  
  lncrna_metaplot_profile <- profile_tss(consensus_peaks[[current_gene]], lncrna_promoters)
  
  mrna_metaplot_profile <- profile_tss(consensus_peaks[[current_gene]], mrna_promoters)
  
  mrna_metaplot_profile$gene_type <- "mRNA"
  lncrna_metaplot_profile$gene_type <- "lncRNA"
  combined_metaplot_profile <- bind_rows(mrna_metaplot_profile, lncrna_metaplot_profile)
  
  ggplot(combined_metaplot_profile, 
         aes(x = x, y = dens, color = gene_type)) +
    geom_vline(xintercept = 0, lty = 2) + 
    geom_line(size = 1.5) + 
    ggtitle( paste( current_gene, " Promoter Metaplot" ) ) +
    scale_x_continuous(breaks = c(-3000, 0, 3000),
                       labels = c("-3kb", "TSS", "+3kb"),
                       name = "") + 
    ylab("Peak frequency") + 
    scale_color_manual(values = c("#424242","#a8404c"))
  ggsave(path="~/CLASS/Justin/CLASS_2021/analysis/06_metaplot/figures/",filename=paste(current_gene,"_both_promoter_metaplot.pdf",sep=""))
}

for (i in 1:length(longer_interest)){
  current_gene <- longer_interest[i]
  all_genes_metaplot_profile <- profile_tss(consensus_peaks[[current_gene]], all_promoters_gr)
  lncrna_metaplot_profile <- profile_tss(consensus_peaks[[current_gene]], lncrna_promoters)
  
  # mRNA promoter profiles
  #mrna_genes <- genes[genes$gene_type == "protein_coding"]
  #mrna_promoters <- promoters(mrna_genes, upstream = 3e3, downstream = 3e3)
  
  #mrna_metaplot_profile <- profile_tss(consensus_peaks[["POLR2A"]], mrna_promoters)
  
  # We can row bind these dataframes so that we can plot them on the same plot
  #mrna_metaplot_profile$gene_type <- "mRNA"
  #lncrna_metaplot_profile$gene_type <- "lncRNA"
  #combined_metaplot_profile <- bind_rows(mrna_metaplot_profile, lncrna_metaplot_profile)
  
  ggplot(all_genes_metaplot_profile, 
         aes(x = x, y = dens)) +
    geom_vline(xintercept = 0, lty = 2) + 
    geom_line(size = 1.5) + 
    ggtitle( paste( current_gene, " Promoter Metaplot" ) ) +
    scale_x_continuous(breaks = c(-3000, 0, 3000),
                       labels = c("-3kb", "TSS", "+3kb"),
                       name = "") + 
    ylab("Peak frequency") + 
    scale_color_manual(values = c("#424242","#a8404c"))
  ggsave(path="~/CLASS/Justin/CLASS_2021/analysis/06_metaplot/figures/",filename=paste(current_gene,"_promoter_metaplot.pdf",sep=""))

}
```

```{r}
promoter_metaplot <- function(peaks, 
                        promoters_gr,
                        upstream = 3e3,
                        downstream = 3e3) {
  

  peak_coverage <- coverage(peaks)
  
  coverage_length <- elementNROWS(peak_coverage)
  coverage_gr <- GRanges(seqnames = names(coverage_length),
                         IRanges(start = rep(1, length(coverage_length)), 
                                 end = coverage_length))
  
  promoters_gr <- subsetByOverlaps(promoters_gr, 
                                       coverage_gr, 
                                       type="within", 
                                       ignore.strand=TRUE)
  chromosomes <- intersect(names(peak_coverage), 
                           unique(as.character(seqnames(promoters_gr))))
  peak_coverage <- peak_coverage[chromosomes]
  
  promoters_ir <- as(promoters_gr, "IntegerRangesList")[chromosomes]
  
  promoter_peak_view <- Views(peak_coverage, promoters_ir)
  
  promoter_peak_view <- lapply(promoter_peak_view, function(x) t(viewApply(x, as.vector)))
  promoter_peak_matrix <- do.call("rbind", promoter_peak_view)
  
  minus_idx <- which(as.character(strand(promoters_gr)) == "-")
  promoter_peak_matrix[minus_idx,] <- promoter_peak_matrix[minus_idx,
                                                           ncol(promoter_peak_matrix):1]
  
  promoter_peak_matrix <- promoter_peak_matrix[rowSums(promoter_peak_matrix) > 1,]
  
  peak_sums <- colSums(promoter_peak_matrix)
  peak_dens <- peak_sums/sum(peak_sums)
  
  metaplot_df <- data.frame(x = -upstream:(downstream-1),
                            dens = peak_dens)
  
  return(metaplot_df)
}
```

