---
title: "02_exploring_data.Rmd"
author: "Katy Walsh"
date: "3/12/2021"
output: html_document
editor_options: 
  chunk_output_type: console
---


```{r setup, include=FALSE}
library(tidyverse)
library(GenomicRanges)
source("util/intersect_functions.R")
source("util/plotting_functions.R")
source("util/_setup.R")
library(Gviz)
#library(ggpubr)
library(ggplot2)
```


load in everything I might need
```{r}
#load promoter annotations
lncrna_promoters <- rtracklayer::import("/scratch/Shares/rinnclass/Katy/CLASS_2021/analysis/00_consensus_peaks/results/gene_annotation/lncrna_promoters.gtf")
mrna_promoters <- rtracklayer::import("/scratch/Shares/rinnclass/Katy/CLASS_2021/analysis/00_consensus_peaks/results/gene_annotation/mrna_promoters.gtf")
lncrna_mrna_promoters <- rtracklayer::import("/scratch/Shares/rinnclass/Katy/CLASS_2021/analysis/00_consensus_peaks/results/gene_annotation/lncrna_mrna_promoters.gtf")

#load gene body annotations
lncrna_gbody <- rtracklayer::import("/scratch/Shares/rinnclass/Katy/CLASS_2021/analysis/00_consensus_peaks/results/gene_annotation/lncrna_genebody.gtf")
mrna_gbody <- rtracklayer::import("/scratch/Shares/rinnclass/Katy/CLASS_2021/analysis/00_consensus_peaks/results/gene_annotation/mrna_genebody.gtf")
lncrna_mrna_gbody <- rtracklayer::import("/scratch/Shares/rinnclass/Katy/CLASS_2021/analysis/00_consensus_peaks/results/gene_annotation/lncrna_mrna_genebody.gtf")


#read in df's
num_peaks_df <- read_csv('analysis/01_global_peak_properties/results/num_peaks_df.csv')
peak_occurence_df <- read_csv('analysis/01_global_peak_properties/results/peak_occurence_dataframe.csv')
lncrna_mrna_promoter_peak_occurence_matrix <- read_table('analysis/01_global_peak_properties/results/lncrna_mrna_promoter_peak_occurence_matrix.tsv')

```

First: looking at # of dbps in lncRNA vs mRNA promoters
```{r}
#using peak_occurence_df plot the graphs with the bimodial distribution from Rmd 10


ggplot(peak_occurence_df, aes(x = number_of_dbp)) + 
  geom_density(alpha = 0.2, color = "#424242", fill = "#424242") +
  theme_paperwhite() +
  xlab(expression("Number of DBPs")) +
  ylab(expression("Density")) +
  ggtitle("Promoter binding events",
          subtitle = "mRNA and lncRNA genes")

#plot in same graph
ggplot(peak_occurence_df, aes(x=number_of_dbp, fill = gene_type, color = gene_type)) +
  geom_density(alpha = 0.0) +
  theme_paperwhite() +
  xlab(expression("Number of DBPs")) +
  ylab(expression("Density")) +
  ggtitle("Promoter binding events")

#save plot (use code below or just use export in the plots window)
# Open a pdf file
pdf("mRNA_lncRNA_separated.pdf") 
# 2. Create a plot
ggplot(peak_occurence_df, aes(x=number_of_dbp, fill = gene_type, color = gene_type)) +
  geom_density(alpha = 0.0) +
  theme_paperwhite() +
  xlab(expression("Number of DBPs")) +
  ylab(expression("Density")) +
  ggtitle("Promoter binding events")
# Close the pdf file
dev.off()

#now we want to remake this for lncRNA or mRNA
#split into separate lncRNA and mRNA promoters
lncrna_peak_occurance <- filter(peak_occurence_df, gene_type == "lncRNA")
mRNA_peak_occurance <- filter(peak_occurence_df, gene_type == "protein_coding")

#plot for just mRNA
ggplot(mRNA_peak_occurance, aes(x=number_of_dbp)) + 
  geom_density(alpha = 0.2, color = "#424242", fill = "#424242") +
  theme_paperwhite() +
  xlab(expression("Number of DBPs")) +
  ylab(expression("Density")) +
  ggtitle("Promoter binding events",
          subtitle = "mRNA promoters")

#plot for just lncRNA
ggplot(lncrna_peak_occurance, aes(x=number_of_dbp)) + 
  geom_density(alpha = 0.2, color = "#424242", fill = "#424242") +
  theme_paperwhite() +
  xlab(expression("Number of DBPs")) +
  ylab(expression("Density")) +
  ggtitle("Promoter binding events",
          subtitle = "lncRNA promoters")

```

Next: are there any DBP's that don't show up in the second hump? (promoters with lots of DBPs bound)
```{r}
##make list of promoters that have lots of DBP's bound (over 200)
#filter genes that have over 200 DBPs
second_peak_promoters <- filter(peak_occurence_df, peak_occurence_df$number_of_dbp > 200)

#make list of promoters that have less than 50 Dbp's bound
first_peak_promoters <- filter(peak_occurence_df, peak_occurence_df$number_of_dbp < 50)

##can run DAVID on those lists of genes to see if there are similar types of genes enriched at either peak
#make csv of gene id's to feed into DAVID
second_peak_gene_ids <- as.data.frame(second_peak_promoters$gene_id)
write_csv(second_peak_gene_names, "analysis/02_exploring_analysis/second_peak_gene_ids")

#same for the first peak
first_peak_gene_ids <- as.data.frame(first_peak_promoters$gene_id)
write_csv(first_peak_gene_ids, "analysis/02_exploring_analysis/first_peak_gene_ids")

#there's 13,000 gene promoters in the second peak and 18,000 in the first so gene ontology analysis likely won't work well since there's so many genes in each
#dplyer::select() would also work instead of as.data.frame()

#could look at if the promoters in our second humps correspond with promoters with lots of DBPs in last class's paper


##now looking at which DBP's ar in each peak and if there are DBPs that aren't in the second peak
#use the matrix that includes each DBP accross each promoter - lncrna_mrna_promoter_peak_matrix
#gene names are in the columns, numbers as the row names (DBPs as first column but as their gene IDs)
matrix_colnames <- colnames(lncrna_mrna_promoter_peak_occurence_matrix)
matrix_rownames1 <- rownames(lncrna_mrna_promoter_peak_occurence_matrix[1])
matrix_column1 <- lncrna_mrna_promoter_peak_occurence_matrix[1]
matrix_column2 <- lncrna_mrna_promoter_peak_occurence_matrix[2]

#want to pull out all columns that are found in the second peak gene id's list... not sure how
#second_peak_promoters_matrix <- 
#    filter(lncrna_mrna_promoter_peak_occurence_matrix,
#            "ENSG00000143870.12" %in% #colnames(lncrna_mrna_promoter_peak_occurence_matrix))
#this isn't working... not sure why. tried to get it to work for one column before matching to second_peak_gene_ids


df_matrix <- as.data.frame(lncrna_mrna_promoter_peak_occurence_matrix)

second_peak_promoters_matrix <- filter(df_matrix, colSums(df_matrix > 200))
#then get row names of filtered table

#re-make the matrix to with just second peak promoters
filtered_consensus_peaks_files <- list.files("/scratch/Shares/rinnclass/Katy/CLASS_2021/analysis/00_consensus_peaks/results/filtered_consensus/", pattern = "*.bed",
                            full.names = TRUE)

filtered_consensus_peaks <- lapply(filtered_consensus_peaks_files, rtracklayer::import)

names(filtered_consensus_peaks) <- gsub("/scratch/Shares/rinnclass/Katy/CLASS_2021/analysis/00_consensus_peaks/results/filtered_consensus//|_filtered_consensus_peaks.bed", "", filtered_consensus_peaks_files)

promoter_peak_occurence <- count_peaks_per_feature(second_peak_gene_ids, filtered_consensus_peaks, type = "occurrence")
#IDs needs to be Granges object




```

```{r}
##where are the uncharacterized zinc finger proteins binding?
#make list of all ZNF proteins
C2H2_ZF_num_peaks_df <- filter(num_peaks_df, dbd == "C2H2 ZF")
C2H2_ZF_names <- as.data.frame(C2H2_ZF_num_peaks_df$dbp)

#make a occurence matrix with just ZNF DBP's
#filter matrix for just ZF DBPs
ZF_matrix <- C2H2_ZF_names %in% df_matrix
#filter didn't work, error variable names are limited to 1000 bytes



```


