---
title: "02_exploring_data.Rmd"
author: "Katy Walsh"
date: "3/12/2021"
output: html_document
editor_options: 
  chunk_output_type: console
---


```{r setup, include=FALSE}
library(tidyverse)
library(GenomicRanges)
source("util/intersect_functions.R")
source("util/plotting_functions.R")
source("util/_setup.R")
library(Gviz)
library(ggpubr)
library(ggplot2)
```


load in everything I might need
```{r}
#load promoter annotations
lncrna_promoters <- rtracklayer::import("/scratch/Shares/rinnclass/Katy/CLASS_2021/analysis/00_consensus_peaks/results/gene_annotation/lncrna_promoters.gtf")
mrna_promoters <- rtracklayer::import("/scratch/Shares/rinnclass/Katy/CLASS_2021/analysis/00_consensus_peaks/results/gene_annotation/mrna_promoters.gtf")
lncrna_mrna_promoters <- rtracklayer::import("/scratch/Shares/rinnclass/Katy/CLASS_2021/analysis/00_consensus_peaks/results/gene_annotation/lncrna_mrna_promoters.gtf")

#load gene body annotations
lncrna_gbody <- rtracklayer::import("/scratch/Shares/rinnclass/Katy/CLASS_2021/analysis/00_consensus_peaks/results/gene_annotation/lncrna_genebody.gtf")
mrna_gbody <- rtracklayer::import("/scratch/Shares/rinnclass/Katy/CLASS_2021/analysis/00_consensus_peaks/results/gene_annotation/mrna_genebody.gtf")
lncrna_mrna_gbody <- rtracklayer::import("/scratch/Shares/rinnclass/Katy/CLASS_2021/analysis/00_consensus_peaks/results/gene_annotation/lncrna_mrna_genebody.gtf")


#read in df's
num_peaks_df <- read_csv('analysis/01_global_peak_properties/results/num_peaks_df.csv')
peak_occurence_df <- read_csv('analysis/01_global_peak_properties/results/peak_occurence_dataframe.csv')
occurence_matrix <- read.table('analysis/01_global_peak_properties/results/lncrna_mrna_promoter_peak_occurence_matrix.tsv')

```

First: looking at # of dbps in lncRNA vs mRNA promoters
```{r}
#using peak_occurence_df plot the graphs with the bimodial distribution from Rmd 10


ggplot(peak_occurence_df, aes(x = number_of_dbp)) + 
  geom_density(alpha = 0.2, color = "#424242", fill = "#424242") +
  theme_paperwhite() +
  xlab(expression("Number of DBPs")) +
  ylab(expression("Density")) +
  ggtitle("Promoter binding events",
          subtitle = "mRNA and lncRNA genes")

#plot in same graph
ggplot(peak_occurence_df, aes(x=number_of_dbp, fill = gene_type, color = gene_type)) +
  geom_density(alpha = 0.0) +
  theme_paperwhite() +
  xlab(expression("Number of DBPs")) +
  ylab(expression("Density")) +
  ggtitle("Promoter binding events")

#save plot (use code below or just use export in the plots window)
# Open a pdf file
pdf("mRNA_lncRNA_separated.pdf") 
# 2. Create a plot
ggplot(peak_occurence_df, aes(x=number_of_dbp, fill = gene_type, color = gene_type)) +
  geom_density(alpha = 0.0) +
  theme_paperwhite() +
  xlab(expression("Number of DBPs")) +
  ylab(expression("Density")) +
  ggtitle("Promoter binding events")
# Close the pdf file
dev.off()

#now we want to remake this for lncRNA or mRNA
#split into separate lncRNA and mRNA promoters
lncrna_peak_occurance <- filter(peak_occurence_df, gene_type == "lncRNA")
mRNA_peak_occurance <- filter(peak_occurence_df, gene_type == "protein_coding")

#plot for just mRNA
ggplot(mRNA_peak_occurance, aes(x=number_of_dbp)) + 
  geom_density(alpha = 0.2, color = "#424242", fill = "#424242") +
  theme_paperwhite() +
  xlab(expression("Number of DBPs")) +
  ylab(expression("Density")) +
  ggtitle("Promoter binding events",
          subtitle = "mRNA promoters")

#plot for just lncRNA
ggplot(lncrna_peak_occurance, aes(x=number_of_dbp)) + 
  geom_density(alpha = 0.2, color = "#424242", fill = "#424242") +
  theme_paperwhite() +
  xlab(expression("Number of DBPs")) +
  ylab(expression("Density")) +
  ggtitle("Promoter binding events",
          subtitle = "lncRNA promoters")



```

Next: are there any DBP's that don't show up in the second hump? (promoters with lots of DBPs bound)
```{r}
##make list of promoters that have lots of DBP's bound (over 200)
#filter genes that have over 200 DBPs
second_peak_promoters <- filter(peak_occurence_df, peak_occurence_df$number_of_dbp > 200)

#make list of promoters that have less than 50 Dbp's bound
first_peak_promoters <- filter(peak_occurence_df, peak_occurence_df$number_of_dbp < 50)

##can run DAVID on those lists of genes to see if there are similar types of genes enriched at either peak
#make csv of gene id's to feed into DAVID
second_peak_gene_ids <- as.data.frame(second_peak_promoters$gene_id)
write_csv(second_peak_gene_names, "analysis/02_exploring_analysis/second_peak_gene_ids")

#same for the first peak
first_peak_gene_ids <- as.data.frame(first_peak_promoters$gene_id)
write_csv(first_peak_gene_ids, "analysis/02_exploring_analysis/first_peak_gene_ids")

#there's 13,000 gene promoters in the second peak and 18,000 in the first so gene ontology analysis likely won't work well since there's so many genes in each
#dplyer::select() would also work instead of as.data.frame()

#could look at if the promoters in our second humps also have 



##now looking at which DBP's ar in each peak and if there are DBPs that aren't in the second peak
#
occurence_matrix <- read.table("results/lncrna_mrna_promoter_peak_occurence_matrix.tsv")

super_promoter <- occurence_matrix[, colSums(occurence_matrix) > 350]
super_promoter_dbps <- data.frame(dbp = rownames(super_promoter), num_super_promoter_overlaps = rowSums(super_promoter))
num_peaks_df <- read.csv("results/num_peaks_df.csv")
#merge the df's will only keep the rows shared between both df's
super_prom_features <- merge(super_promoter_dbps, num_peaks_df)
super_promoter_features <- dplyr::select(super_promoter_features, dbp, num_super_promoter_overlaps, everything() )
super_promoter_features <- super_promoter_features %>%
  mutate(percent_super_prom_ov = num_super_promoter_overlaps/peaks_overlapping_promoters * 100, 
         percent_of_total_peaks_at_super_prom = num_super_promoter_overlaps/num_peaks * 100,
         zincfinger = ifelse(grepl(" ZF",dbd), "ZF", "Other"))
table(super_promoter_features$zincfinger)
ggplot(super_promoter_features, aes(x = percent_super_prom_ov, color = zincfinger)) +
  geom_density()
ggplot(super_promoter_features, aes(x = percent_of_total_peaks_at_super_prom, y = percent_super_prom_ov)) + 
  geom_point()
write.csv(super_promoter_features, "results/super_promoter_features.csv")



#conclusion: it doesn't look like the ZF's are binding at higer rates to the "super promoters" than other DBPs
```

```{r}
##where are the uncharacterized zinc finger proteins binding?
#make list of all ZNF proteins

#make a occurence matrix with just ZNF DBP's


```


