---
title: "RNA_seq_fractions"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
options(stringsAsFactors = FALSE)
library(tidyverse)
library(tximport)
library(DESeq2)
library(ComplexHeatmap)
library(circlize)
library(ggdendro)
library(pheatmap)

source("../../util/plotting_functions.R")
source("../../util/_setup.R")
```

Read in the count data.

```{r}
samplesheet <- read_csv("/scratch/Shares/rinnclass/JR/CLASS_2021/rnaseq/samplesheet.csv")

# The cleanest thing to do would be to fix it on the design file and re-run the nf-core pipeline,
# but for now we can swap it here.

samplesheet[which(samplesheet$sample_name == "hepg2_R1"), "condition"] <- "insoluble_cytoplasmic_fraction"
samplesheet[which(samplesheet$sample_name == "hepg2_insoluble_cytoplasmic_fraction_R2"), "condition"] <- "total"
#write_csv(sample_sheet,"/scratch/Shares/rinnclass/data/samplesheet_fixed_condition.csv")

```



```{r}
samplesheet <- read_csv("/scratch/Shares/rinnclass/data/samplesheet_fixed_condition.csv")

# We want to select just a subset of these samples
# so that we can run differential expression for two conditions
samplesheet <- samplesheet %>%
  filter(condition %in% c("total", "nuclear_fraction"))
```

```{r}
# We want to include the gene annotation info in the DEseq2 object,
# so we'll need to read that in.
gencode_gtf <- rtracklayer::import("/scratch/Shares/rinnclass/data/genomes/gencode.v32.annotation.gtf")
genes <- gencode_gtf[gencode_gtf$type == "gene"]

# Salmon's primary level of quantifying is at the transcript-level.
# Since we're doing differential expression at the gene-level
# Salmon needs to be able to map transcripts to genes.
tx2gene <- gencode_gtf[gencode_gtf$type == "transcript"] %>% 
  as.data.frame() %>%
  dplyr::select(transcript_id, gene_id)

# We'll also create an object that maps gene IDs to gene symbols
# This will come in handy in looking at our results
g2s <- genes %>% as.data.frame() %>%
  dplyr::select(gene_id, gene_name)

```

```{r}
# Read in the count data from salmon
# For DESeq2 we want the raw counts
# The easiest way to import it in the proper format is to 
# use the tximport package
files <- file.path("/scratch/Shares/rinnclass/Nina/CLASS_2021/rnaseq/results/star_salmon",
                   samplesheet$sample_name, 
                   "quant.sf")
names(files) <- samplesheet$sample_name

txi <- tximport(files, type = c("salmon"), tx2gene = tx2gene)

```


```{r}
# Will be adding the gene information, the count data, and the sample information
# to the DESeq function. We want to ensure that all of this data is synced up
# So we need to reorder the data so that it's all matching.
names(genes) <- genes$gene_id
genes <- genes[rownames(txi$counts)]
stopifnot(all(genes$gene_id == rownames(txi$counts)))

# Create rownames for the sample sheet and then use that to re-order
samplesheet <- samplesheet %>%
  mutate(row_names = sample_name) %>%
  column_to_rownames("row_names")
samplesheet <- samplesheet[colnames(txi$counts),]
stopifnot(all(rownames(samplesheet) == colnames(txi$counts)))

```


```{r}
# The conditions that we'll be comparing here are 
# are nuclear and cytoplasmic fractions
# Normally with WT and knockout experiments, you'd
# compare the treatment (KO) condition back to the control
# condition (WT)
# Here we don't really have an obvious choice of "control"
# So let's just choose to compare back to nuclear as our control
# To do that we make it the first level of the factor
samplesheet$condition <- factor(samplesheet$condition, levels = c("total", "nuclear_fraction"))

# Now we have all the components that we need to add to the 
# DESeq2 function. So now let's create our DEseq2 object
#set.seed(789723)
dds <- DESeqDataSetFromTximport(txi,
                                design = ~ condition,
                                colData = samplesheet,
                                rowData = genes)
```

```{r}
# We now have the object set up to 
# run the differential expression analysis
# and it's a very simple command to do so. 
# Even though it's doing a lot of complicated statistics!
dds <- DESeq(dds)
```

```{r}
# The first thing that we'll want to do is make a PCA plot.
rlog_counts <- rlog(dds)


rld_pca <- prcomp(t(assay(rlog_counts)))
rld_prcomps <- rld_pca$x %>% as.data.frame() %>%
  rownames_to_column("sample_name") %>%
  select(sample_name, PC1, PC2)
rld_prcomps <- merge(samplesheet, rld_prcomps)
g <- ggplot(rld_prcomps, aes(x = PC1, y = PC2, color = condition))
g + geom_point() 
```


