---
title: "Global peak properties"
output: html_document
editor_options: 
  chunk_output_type: console
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library(tidyverse)
library(GenomicRanges)
library(Gviz)
library(ggpubr)
library(BiocManager)
    install("BSgenome.Hsapiens.UCSC.hg38")
BiocManager::install("biomartr")
library(biomartr)
library(regioneR)
library(ggdendro)
    
source("/scratch/Shares/rinnclass/Justin/util/intersect_functions.R")
source("/scratch/Shares/rinnclass/Justin/CLASS_2021/util/plotting_functions.R")
source("/scratch/Shares/rinnclass/Justin/CLASS_2021/util/_setup.R")
```

```{r}
base_path <- "/scratch/Shares/rinnclass/Justin/CLASS_2021/analysis/00_consensus_peaks/results"
peak_list <- import_peaks(file.path(base_path,
                                    "filtered_consensus/"))
# Import gencode
lncrna_promoters <- rtracklayer::import(file.path(base_path,
                                                  "promoters/lncrna_promoters.gtf"))
mrna_promoters <- rtracklayer::import(file.path(base_path,
                                                "promoters/mrna_promoters.gtf"))
lncrna_mrna_promoters <- rtracklayer::import(file.path(base_path,
                                                "promoters/lncrna_mrna_promoters.gtf"))
lncrna_matrix <- count_peaks_per_feature(lncrna_promoters, 
                                         peak_list, type = "occurrence")
mrna_matrix <- count_peaks_per_feature(mrna_promoters, 
                                       peak_list, type = "occurrence")

```

```{r}
num_peaks_df <- data.frame("dbp" = rownames(lncrna_matrix),
                           "total_number_of_peaks" = sapply(peak_list, length),
                           "lncrna" = rowSums(lncrna_matrix),
                           "mrna" = rowSums(mrna_matrix))

summary(lncrna_matrix)
```

Find unbound and enough-bounding promoters
```{r} 
unbound_promoters <- peak_occurence_df %>% filter(peak_occurence_df$number_of_dbp < 1)

```

Find DBPs with no overlaps with promoter
```{r}
dbp_promoter_ovl <- get_overlapping_peaks(lncrna_mrna_promoters, filtered_consensus_peaks)

head(dbp_promoter_ovl)
num_ovl <- sapply(dbp_promoter_ovl, length)
min(num_ovl)
summary(num_ovl)

```

Plot peak_occurence_df
```{r}
g <- ggplot(peak_occurence_df, aes(x = number_of_dbp))

#AES layer will know to make a histogram since there is only one set of values for x.
g + geom_density(alpha = 0.2, color = "#424242", fill = "#424242") +
  xlab(expression("Number of DBPs")) +
  ylab(expression("Density")) +
  ggtitle("DBPs distribution",
          subtitle = "mRNA and lncRNA genes") 

```

Dive deeper into promoters with num of DBPs >= 170
```{r}
high_bound_promoters <- peak_occurence_df %>% filter(peak_occurence_df$number_of_dbp >= 200, peak_occurence_df$number_of_dbp <= 350)

ggplot(high_bound_promoters, aes(
      x = gene_name,
      y = number_of_dbp)) +
        geom_bar(stat = "identity")

high_bound_id <- high_bound_promoters$gene_id 

filtered_promoters <- lncrna_mrna_promoters[lncrna_mrna_promoters$gene_id %in% high_bound_id]
#length(filtered_promoters)
high_bound_occur <- count_peaks_per_feature(filtered_promoters, peak_list, 
                                               type = "occurrence")

filtered_promoters[1]
features_names <- filtered_promoters$gene_name
peak_names <- names(peak_list)
temp <- matrix(0, ncol = length(filtered_promoters), nrow = length(peak_list))
rownames(temp) <- peak_names
colnames(temp) <- features_names
temp[1,features_names[1]]

source("/scratch/Shares/rinnclass/Justin/util/intersect_functions.R")
peak_store <- get_overlapping_peaks(filtered_promoters, peak_list, 1)
peak_store[peak_names[1],features_names[10]]

high_bound_df <- as.data.frame(as.table(peak_store))
write_csv(high_bound_df, "/scratch/Shares/rinnclass/Justin/CLASS_2021/analysis/02_content_of_peaks_per_promoter/results/interesting_feature_peaks_dataframe.csv")

```
Permutation Test
```{r}
high_bound_df
hg38 <- getGenome("hg38")
suppressWarnings(pt <- overlapPermTest(A = filtered_promoters, 
                                       B = peak_list, 
                                       ntimes = 1000, 
                                       non.overlapping = FALSE, 
                                       verbose = TRUE,
                                       genome = hg38,
                                       alternative =  "auto", 
                                       mc.cores = 8))


```
RowSum, ColumnSum of peak_store matrix
```{r}
po_2020_df <- read.csv(file = '/scratch/Shares/rinnclass/data/2020_peak_occurence_df.csv')
dbp_promoter_binding_table <- read.csv(file = '/scratch/Shares/rinnclass/data/dbp_promoter_binding_table.csv')
```

h_clustering
```{r}
bin_hier <- hclust(dist(high_bound_occur, method = "binary"))
bin_hier$labels[bin_hier$order]
ggdendro::ggdendrogram(bin_hier, rotate = FALSE,  size = 3, 
                       theme_dendro = TRUE) +
  coord_flip() +
  # scale_y_continuous() +
  # scale_x_continuous(position = "top") +
  scale_x_continuous(breaks = seq_along(bin_hier$labels[bin_hier$order]),
            labels = bin_hier$labels[bin_hier$order], position = "top",
            expand = c(0,0)) +
  theme(axis.text.x = element_text(angle = 90, hjust  = 1)) + 
  theme(axis.text.y = element_text(angle = 0,hjust = 1)) +
  scale_y_reverse(expand = c(0.01, 0)) +
  theme(
    plot.background = element_blank(),
    panel.grid.major = element_blank(),
    panel.grid.minor = element_blank(),
    panel.border = element_blank()
  )
ggsave("figures/all_hclust_binary_dist.pdf", height = 22, width = 2.5)
ggsave("figures/all_hclust_binary_dist.png", height = 22, width = 2.5)
```

Scatter plot of percentage of reservoirs in K562 and those in HEPG2
```{r}
po2021_df <- peak_occurence_df
#get the reservoirs in 2020 peak_occurence_df
reservoir_2020 <- po_2020_df %>% filter(po_2020_df$reservoir == 1) 
reservoir_name_list <- reservoir_2020$gene_name
#get the reserviors in 2021 peak_occurence_df
reservoir_2021 <- po2021_df %>% filter(po2021_df$gene_name %in% reservoir_name_list)
total_sum_2021 <- sum(po2021_df$number_of_dbp)
total_sum_2020 <- sum(po2020_df$number_of_dbp)
#normalize by the total number of DBP
normalized_2021 <- peak_occurence_df %>%
  dplyr::select(gene_id, gene_type, number_of_dbp) %>%
  dplyr::mutate(dbp_norm = number_of_dbp/sum()) %>%
  dplyr::rename(norm_num_dpb_HEPG2 = dbp_norm)


normalized_reservoir_2021 <- reservoir_2021 %>%
  dplyr::select(gene_id, gene_type, number_of_dbp) %>%
  dplyr::mutate(dbp_norm = number_of_dbp/sum()) %>%
  dplyr::rename(norm_num_dpb_reservoir = dbp_norm)

normalized_reservoir_2021 <- reservoir_2021 %>%
  dplyr::select(gene_id, gene_type, number_of_dbp) %>%
  dplyr::mutate(dbp_norm = number_of_dbp/length(peak_names)) %>%
  dplyr::rename(norm_num_dpb_reservoir = dbp_norm)
comparison_df <- merge(normalized_2021, num_reservoir_2021)

summary(dbp_promoter_binding_table)
ggplot(comparison_df, aes(x = num_dbp_reservoir, 
                 y = num_dpb_HEPG2)) +
  geom_point()
```


