---
title: "02_exploring_data.Rmd"
author: "Katy Walsh"
date: "3/12/2021"
output: html_document
editor_options: 
  chunk_output_type: console
---


```{r setup, include=FALSE}
library(tidyverse)
library(GenomicRanges)
source("util/intersect_functions.R")
source("util/plotting_functions.R")
source("util/_setup.R")
library(Gviz)
library(ggpubr)
library(ggplot2)
```


load in everything I might need
```{r}
#load promoter annotations
lncrna_promoters <- rtracklayer::import("/scratch/Shares/rinnclass/Katy/CLASS_2021/analysis/00_consensus_peaks/results/gene_annotation/lncrna_promoters.gtf")
mrna_promoters <- rtracklayer::import("/scratch/Shares/rinnclass/Katy/CLASS_2021/analysis/00_consensus_peaks/results/gene_annotation/mrna_promoters.gtf")
lncrna_mrna_promoters <- rtracklayer::import("/scratch/Shares/rinnclass/Katy/CLASS_2021/analysis/00_consensus_peaks/results/gene_annotation/lncrna_mrna_promoters.gtf")

#load gene body annotations
lncrna_gbody <- rtracklayer::import("/scratch/Shares/rinnclass/Katy/CLASS_2021/analysis/00_consensus_peaks/results/gene_annotation/lncrna_genebody.gtf")
mrna_gbody <- rtracklayer::import("/scratch/Shares/rinnclass/Katy/CLASS_2021/analysis/00_consensus_peaks/results/gene_annotation/mrna_genebody.gtf")
lncrna_mrna_gbody <- rtracklayer::import("/scratch/Shares/rinnclass/Katy/CLASS_2021/analysis/00_consensus_peaks/results/gene_annotation/lncrna_mrna_genebody.gtf")


#read in df's
num_peaks_df <- read_csv('analysis/01_global_peak_properties/results/num_peaks_df.csv')
peak_occurence_df <- read_csv('analysis/01_global_peak_properties/results/peak_occurence_dataframe.csv')
lncrna_mrna_promoter_peak_occurence_matrix <- read_table('analysis/01_global_peak_properties/results/lncrna_mrna_promoter_peak_occurence_matrix.tsv')

```

First: looking at # of dbps in lncRNA vs mRNA promoters
```{r}
#using peak_occurence_df plot the graphs with the bimodial distribution from Rmd 10


ggplot(peak_occurence_df, aes(x = number_of_dbp)) + 
  geom_density(alpha = 0.2, color = "#424242", fill = "#424242") +
  theme_paperwhite() +
  xlab(expression("Number of DBPs")) +
  ylab(expression("Density")) +
  ggtitle("Promoter binding events",
          subtitle = "mRNA and lncRNA genes")

#plot in same graph
ggplot(peak_occurence_df, aes(x=number_of_dbp, fill = gene_type, color = gene_type)) +
  geom_density(alpha = 0.0) +
  theme_paperwhite() +
  xlab(expression("Number of DBPs")) +
  ylab(expression("Density")) +
  ggtitle("Promoter binding events")

#save plot (use code below or just use export in the plots window)
# Open a pdf file
pdf("mRNA_lncRNA_separated.pdf") 
# 2. Create a plot
ggplot(peak_occurence_df, aes(x=number_of_dbp, fill = gene_type, color = gene_type)) +
  geom_density(alpha = 0.0) +
  theme_paperwhite() +
  xlab(expression("Number of DBPs")) +
  ylab(expression("Density")) +
  ggtitle("Promoter binding events")
# Close the pdf file
dev.off()

#now we want to remake this for lncRNA or mRNA
#split into separate lncRNA and mRNA promoters
lncrna_peak_occurance <- filter(peak_occurence_df, gene_type == "lncRNA")
mRNA_peak_occurance <- filter(peak_occurence_df, gene_type == "protein_coding")

#plot for just mRNA
ggplot(mRNA_peak_occurance, aes(x=number_of_dbp)) + 
  geom_density(alpha = 0.2, color = "#424242", fill = "#424242") +
  theme_paperwhite() +
  xlab(expression("Number of DBPs")) +
  ylab(expression("Density")) +
  ggtitle("Promoter binding events",
          subtitle = "mRNA promoters")

#plot for just lncRNA
ggplot(lncrna_peak_occurance, aes(x=number_of_dbp)) + 
  geom_density(alpha = 0.2, color = "#424242", fill = "#424242") +
  theme_paperwhite() +
  xlab(expression("Number of DBPs")) +
  ylab(expression("Density")) +
  ggtitle("Promoter binding events",
          subtitle = "lncRNA promoters")




```


