---
title: "12_Clustering"
output: html_document
editor_options: 
  chunk_output_type: console
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library(magrittr)
library(tidyverse)
library(ggdendro)
library(GenomicRanges)
source("util/intersect_functions.R")
source("util/plotting_functions.R")
source("util/_setup.R")
```


```{r}
peak_occurence_matrix <- read.table("~/CLASS/Justin/CLASS_2021/analysis/01_global_peak_properties/results/lncrna_mrna_promoter_peak_occurence_matrix.tsv")

peak_occurence_matrix <- peak_occurence_matrix[rowSums(peak_occurence_matrix) > 250, ]

peak_occurence_dist <- dist(peak_occurence_matrix, method = "binary")


# Hierarchical clustering with binary distance measure
bin_hier <- hclust(peak_occurence_dist, method = "complete")

plot(bin_hier)

# To make this bigger, we can plot it as a pdf
pdf("figures/dbp_hclust_dendro.pdf", height = 12, width = 70)
plot(bin_hier)
dev.off()

bin_hier$labels[bin_hier$order]
```


```{r}

# # Let's start with lncRNA promoters -- why not?
# 
# # We first need to load the .GTF we made for lncRNAs and mRNAs.
# 
# lncrna_promoters <- rtracklayer::import("results/lncrna_promoters.gtf")
# mrna_promoters <- rtracklayer::import("results/mrna_promoters.gtf")
# 
# # Cool now we can do the same as above:
# 
# lncrna_peak_occurence <- peak_occurence_matrix[,lncrna_promoters$gene_id]
# 
# bin_hier_lncrna <- hclust(dist(lncrna_peak_occurence, method = "binary"))
# 
# ggdendro::ggdendrogram(bin_hier_lncrna, rotate = TRUE,  size = 3)
# 
# ggsave("chipseq/figures/lncrna_hclust_binary_dist.pdf", height = 44, width = 6)
# 
# # Cool we can see the Pol II S2 and S5 diff right away again!
# 
# 
# # Now for mRNA
# 
# mrna_peak_occurence <- peak_occurence_matrix[,mrna_promoters$gene_id]
# 
# bin_hier_mrna <- hclust(dist(lncrna_peak_occurence, method = "binary"))
# 
# ggdendro::ggdendrogram(bin_hier, rotate = TRUE,  size = 3)
# 
# ggsave("chipseq/figures/mrna_hclust_binary_dist.pdf", height = 44, width = 6)


# Do a file transfer and open up lncRNA and mRNA clusters side by side:
# They are almost identical !! That is another good sign.

# Nice I think we have a solid data set here ! Let's explore more in the next class.

```

Class excercise: find regions of the clusters that make sense. For example my favorite is that SETDB1 is the methyltransferase that modifies histones to have H3K9me3. Note the two cluster next to each other -- another good sign.

Let's make a heatmap of the promoters -- since it will take a long long time to plot we'll just plot the "super promoters"


```{r}
# We'll start with the promoter peak occurrence and filter down to just those that have high numbers of dbps

genes_of_interest <- c("ZZZ3","ZNF343","ZNF34","ATF5")

clust11_shared_promoters <- peak_occurence_matrix[rownames(peak_occurence_matrix) %in% longer_interest,colSums(peak_occurence_matrix) >= length(longer_interest)]

clust11_shared_promoters_df <- data.frame(gene_id = colnames(clust11_shared_promoters))
clust11_shared_promoters_df <- strsplit(as.character(clust11_shared_promoters_df), ".", fixed = TRUE) %>% sapply(extract2, 1)
write(clust11_shared_promoters_df,"~/CLASS/Justin/CLASS_2021/analysis/04_global_clustering/results/cluster11_shared_promoters.txt")

shared_df <- data.frame(gene_id = colnames(shared_promoters))
shared_df <- strsplit(as.character(shared_df$gene_id), ".", fixed = TRUE) %>% sapply(extract2, 1)
write(shared_df,"~/CLASS/Justin/CLASS_2021/analysis/04_global_clustering/results/shared_promoters.txt")

ZZZ3_promoters_list <- as.data.frame(t(ZZZ3_promoters))
ZZZ3_df <- po_2021_df[as.character(rownames(ZZZ3_promoters_list)),]


po_2021_df <- read.csv(file = '~/CLASS/Justin/CLASS_2021/analysis/01_global_peak_properties/results/peak_occurence_dataframe.csv')
list <- po_2021_df[ZZZ3_promoters_list$id,]
promoter_list <- as.character(list$gene_name)
write(promoter_list,"~/CLASS/Justin/CLASS_2021/analysis/04_global_clustering/results/ZZZ3_promoters.txt")
zinc_high_binders <- high_binders[zinc_finger_dbd$dbp,]
min(rowSums(high_binders))

# If you don't have pheatmap, install it.


library(pheatmap)

# We don't need to be able to read the promoter names and it takes a long time to plot those.
pheatmap(high_binders, show_colnames = FALSE, clustering_distance_rows = "binary", clustering_distance_cols = "binary")

po_hclust <- hclust(dist(peak_occurence_matrix, method = "binary"))
plot(po_hclust)


# If we want to just plot the zinc fingers for example, we can reference the num_peaks_df 
num_peaks_df <- read_csv("~/CLASS/Justin/CLASS_2021/analysis/03_reservoir_binding_DBP/results/super_promoter_features.csv")
num_peaks_df$X1 <- NULL
zinc_finger_dbd <- num_peaks_df %>% filter(zincfinger == "ZF")

# Index this matrix using the names
zf_matrix <- high_binders[zinc_finger_dbd$dbp,] 
compare_matrix <- peak_occurence_matrix[comparison$dbp,colSums(peak_occurence_matrix) > 350]
pheatmap(zf_matrix, show_colnames = FALSE, clustering_distance_rows = "binary", clustering_distance_cols = "binary")
pheatmap(compare_matrix, show_colnames = FALSE, clustering_distance_rows = "binary", clustering_distance_cols = "binary")

# We can also check out the dendrogram for just this matrix.

zf_hclust <- hclust(dist(zf_matrix, method = "binary"))
plot(zf_hclust)
pdf("figures/zf_hclust_dendro.pdf", height = 12, width = 70)

compare_hclust <- hclust(dist(compare_matrix, method = "binary"))
plot(zf_hclust)
```

```{r}
clust <- cutree(bin_hier,h=.52)
table(clust)
clust_df <- data.frame(dbp = names(clust), cluster = clust)
#choos dbps in cluster11
cluster11 <- filter(clust_df,cluster == 11)
longer_interest <- as.character(cluster11$dbp)
#peak occurrences of cluster 11 dbps
po_clust11 <- peak_occurence_matrix[row.names(peak_occurence_matrix) %in% row.names(cluster11),]
#promoters that ALL dpbs in cluster 11 bind (there are 10 dpbs)
pro_cluster11 <- po_clust11[,colSums(po_clust11) > 9]
#pull out the promoter names that all dbps in cluster11 bound
pro_clus11 <- data.frame(promoters = colnames(pro_cluster11))
```

